<!DOCTYPE html>
<html lang="en">
<head>
    

    




    

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <title>Manual Pre v1.1 | HISAT-genotype</title>

    <meta name="description" content="next-generation genomic analysis software platform">

    <link rel="shortcut icon" href="/hisat-genotype/assets/img/ogp.png" type="image/x-icon">
    <link rel="canonical" href="http://localhost:4000/hisat-genotype/manual_old/">
    <link rel="alternate" type="application/atom+xml" title="HISAT-genotype" href="/hisat-genotype/feed.xml" />

    

    <meta property="og:title" content="Manual Pre v1.1 | HISAT-genotype" />
    <meta property="og:type" content="website" />
    <meta property="og:image" content="//ttskch.github.io/jekyll-ttskch-theme/assets/img/ogp.png" />
    <meta property="og:url" content="http://localhost:4000/hisat-genotype/manual_old/" />
    <meta property="og:site_name" content="HISAT-genotype" />
    <meta property="fb:admins" content="" />
    <meta property="fb:app_id" content="" />
    <meta property="og:description" content="next-generation genomic analysis software platform" />

    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
    <script src="https://use.fontawesome.com/1f5f360d80.js"></script>
    <link href="//fonts.googleapis.com/css?family=Source+Sans+Pro:400,700,700italic,400italic" rel="stylesheet">

    <link href="/hisat-genotype/assets/css/style.css" rel="stylesheet">
</head>
<body>

<header class="site-header">
    <div class="inner clearfix">
        
        <a href="/hisat-genotype/">
            <img class="avatar" src="/hisat-genotype/assets/img/ogp.png" alt=""/>
        </a>
        
        <h1 class="clearfix">
            <a class="title " href="/hisat-genotype/">HISAT-genotype</a>
            <br><span class="description">next-generation genomic analysis software platform</span>
        </h1>
    </div>
</header>

<div class="site-container">
    <div class="site-content">
        <div class="article-wrapper">
    <article>
        

<header>
    <div class="panel">
        <h1>
            
            Manual Pre v1.1
            
        </h1>

        <ul class="tags">
            
            
            
        </ul>

        <div class="clearfix">
            <ul class="meta">
                

                
            </ul>
        </div>
    </div>

    

    
</header>

        <section class="post-content">
            
<ul id="markdown-toc">
  <li><a href="#introduction" id="markdown-toc-introduction">Introduction</a></li>
  <li><a href="#scripts" id="markdown-toc-scripts">Scripts</a>    <ul>
      <li><a href="#hisatgenotypepy--analysis-of-a-whole-human-genome" id="markdown-toc-hisatgenotypepy--analysis-of-a-whole-human-genome">hisatgenotype.py – Analysis of a whole human genome</a></li>
      <li><a href="#hisatgenotype_locuspy--analysis-of-a-specific-gene-genomic-region-or-gene-family" id="markdown-toc-hisatgenotype_locuspy--analysis-of-a-specific-gene-genomic-region-or-gene-family">hisatgenotype_locus.py – Analysis of a specific gene, genomic region, or gene family</a>        <ul>
          <li><a href="#for-advanced-users" id="markdown-toc-for-advanced-users">For advanced users</a></li>
        </ul>
      </li>
      <li><a href="#hisatgenotype_build_genomepy--building-a-genotype-genome" id="markdown-toc-hisatgenotype_build_genomepy--building-a-genotype-genome">hisatgenotype_build_genome.py – Building a genotype genome</a></li>
      <li><a href="#hisatgenotype_extract_varspy--extraction-of-variants-haplotypes-etc" id="markdown-toc-hisatgenotype_extract_varspy--extraction-of-variants-haplotypes-etc">hisatgenotype_extract_vars.py – Extraction of variants, haplotypes, etc.</a></li>
      <li><a href="#hisatgenotype_extract_readspy--extraction-of-reads-that-belong-to-genomic-regions-of-interest" id="markdown-toc-hisatgenotype_extract_readspy--extraction-of-reads-that-belong-to-genomic-regions-of-interest">hisatgenotype_extract_reads.py – Extraction of reads that belong to genomic regions of interest</a></li>
      <li><a href="#hisatgenotype_locus_samplespy--analysis-of-many-samples" id="markdown-toc-hisatgenotype_locus_samplespy--analysis-of-many-samples">hisatgenotype_locus_samples.py – Analysis of many samples</a></li>
    </ul>
  </li>
</ul>

<h1 id="introduction">Introduction</h1>

<p>This is the old manual for versions earlier than 1.1. Please use this information for those versions only.</p>

<hr />

<h1 id="scripts">Scripts</h1>

<h2 id="hisatgenotypepy--analysis-of-a-whole-human-genome">hisatgenotype.py – Analysis of a whole human genome</h2>
<p>The hisatgenotype.py python script will analyze a whole human genome using whole genome sequencing reads. It will align reads to genotype genome, extract reads belonging to each locus of interest, and perform typing and assembly.</p>

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Description</th>
      <th>Default</th>
      <th>Example</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>--base</strong></td>
      <td>Base file name for index, variants, haplotypes, etc.</td>
      <td>genotype_genome</td>
      <td><code class="highlighter-rouge">--base genotype_genome</code></td>
    </tr>
    <tr>
      <td><strong>--region-list</strong></td>
      <td>Comma separated list of regions</td>
      <td>(empty) meaning every region available</td>
      <td><code class="highlighter-rouge">--region-list hla.A,hla.B,hla.C,codis.FGA,cyp.2D6</code></td>
    </tr>
    <tr>
      <td><strong>-p / --threads</strong></td>
      <td>Number of threads to be used</td>
      <td>1</td>
      <td><code class="highlighter-rouge">-p 4</code></td>
    </tr>
    <tr>
      <td><strong>-U</strong></td>
      <td>Single-end read file name</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-U read.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-1</strong></td>
      <td>Paired-end read file name 1</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-1 read.1.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-2</strong></td>
      <td>Paired-end read file name 2</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-2 read.2.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-f / --fasta</strong></td>
      <td>Reads are provided in FASTA format</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-f</code></td>
    </tr>
    <tr>
      <td><strong>--alignment-file</strong></td>
      <td>Sorted bam alignment file converted from HISAT2’s sam alignment output</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">--alignment-file NA12878.bam</code></td>
    </tr>
    <tr>
      <td><strong>--num-editdist</strong></td>
      <td>Number of maximum edit distance or mismatches allowed in the alignment, typing, and assembly of a read (not a pair)</td>
      <td>2</td>
      <td><code class="highlighter-rouge">--num-editdist 0</code></td>
    </tr>
    <tr>
      <td><strong>--assembly</strong></td>
      <td>Perform assembly of each locus of interest</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--assembly</code></td>
    </tr>
    <tr>
      <td><strong>--verbose</strong></td>
      <td>Provide more information</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--verbose</code></td>
    </tr>
    <tr>
      <td><strong>-h / --help</strong></td>
      <td>Output help message</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--help </code></td>
    </tr>
  </tbody>
</table>

<p>Examples:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>hisatgenotype.py <span class="nt">--base</span> genotype_genome <span class="nt">-p</span> 4 <span class="nt">-1</span> read.1.fq.gz <span class="nt">-2</span> read.2.fq.gz
<span class="nv">$ </span>hisatgenotype.py <span class="nt">--verbose</span> <span class="nt">--base</span> genotype_genome <span class="nt">-p</span> 4 <span class="nt">--region-list</span> hla.A <span class="nt">-1</span> read.1.fq.gz <span class="nt">-2</span> read.2.fq.gz
<span class="nv">$ </span>hisatgenotype.py <span class="nt">--base</span> genotype_genome <span class="nt">--alignment-file</span> NA12878.bam  
<span class="nv">$ </span>hisatgenotype.py <span class="nt">--base</span> genotype_genome <span class="nt">--region-list</span> hla.A,hla.B,hla.C,hla.DQA1,hla.DQB1,hla.DRB1,codis,cyp.2D6 <span class="nt">--alignment-file</span> NA12878.bam
</code></pre></div></div>

<h2 id="hisatgenotype_locuspy--analysis-of-a-specific-gene-genomic-region-or-gene-family">hisatgenotype_locus.py – Analysis of a specific gene, genomic region, or gene family</h2>
<p>The hisatgenotype_locus.py python script will analyze a particular locus or a set of genes or genomic regions using extracted reads. As the name of the script (locus) implies, it will align reads to a part of genotype genome and perform typing and assembly.</p>

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Description</th>
      <th>Default</th>
      <th>Example</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>--base</strong></td>
      <td>Base file name for index, variants, haplotypes, etc.</td>
      <td>hla</td>
      <td><code class="highlighter-rouge">--base codis</code></td>
    </tr>
    <tr>
      <td><strong>--locus-list</strong></td>
      <td>Comma separated list of loci</td>
      <td>(empty) meaning every locus available</td>
      <td><code class="highlighter-rouge">--locus-list A,B,C,DQA1,DQB1,DRB1</code></td>
    </tr>
    <tr>
      <td><strong>-p / --threads</strong></td>
      <td>Number of threads to be used</td>
      <td>1</td>
      <td><code class="highlighter-rouge">-p 4</code></td>
    </tr>
    <tr>
      <td><strong>-U</strong></td>
      <td>Single-end read file name</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-U read.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-1</strong></td>
      <td>Paired-end read file name 1</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-1 read.1.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-2</strong></td>
      <td>Paired-end read file name 2</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-2 read.2.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-f / --fasta</strong></td>
      <td>Reads are provided in FASTA format</td>
      <td>false</td>
      <td><code class="highlighter-rouge">-f</code></td>
    </tr>
    <tr>
      <td><strong>--num-editdist</strong></td>
      <td>Number of maximum edit distance or mismatches allowed in the alignment, typing, and assembly of a read (not a pair)</td>
      <td>2</td>
      <td><code class="highlighter-rouge">--num-editdist 0</code></td>
    </tr>
    <tr>
      <td><strong>--assembly</strong></td>
      <td>Perform assembly of each locus of interest</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--assembly</code></td>
    </tr>
    <tr>
      <td><strong>--genotype-genome</strong></td>
      <td>Base name for genotype genome, which the program will use instead of region-based small indexes</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">--genotype-genome genotype_genome</code></td>
    </tr>
    <tr>
      <td><strong>--verbose</strong></td>
      <td>Provide more information</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--verbose</code></td>
    </tr>
    <tr>
      <td><strong>-h / --help</strong></td>
      <td>Output help message</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--help</code></td>
    </tr>
  </tbody>
</table>

<p>Examples:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="nv">$ </span>hisatgenotype_locus.py <span class="nt">--base</span> hla <span class="nt">--locus-list</span> A <span class="nt">--num-editdist</span> 2 <span class="nt">--assembly</span> <span class="nt">-1</span> ILMN/NA12892.extracted.1.fq.gz <span class="nt">-2</span> ILMN/NA12892.extracted.2.fq.gz
 <span class="nv">$ </span>hisatgenotype_locus.py <span class="nt">--base</span> hla <span class="nt">--locus-list</span> A,B,C,DQA1,DQB1,DRB1 <span class="nt">--num-editdist</span> 2 <span class="nt">-1</span> ILMN/NA12892.extracted.1.fq.gz <span class="nt">-2</span> ILMN/NA12892.extracted.2.fq.gz
</code></pre></div></div>

<h3 id="for-advanced-users">For advanced users</h3>
<p>In addition to analyzing real sequencing reads, the script comes with additional testing capabilities, which might come in handy when determining whether a problematic result is due to noise in sequencing reads or algorithms. To do so, the script can generate simulation reads with sequencing errors and SNPs, and test and evaluate the accuracy of its typing and assembly algorithms, with additional options described below.</p>

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Description</th>
      <th>Default</th>
      <th>Example</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>--read-len</strong></td>
      <td>Length of a simulated read</td>
      <td>100</td>
      <td><code class="highlighter-rouge">--read-len 150</code></td>
    </tr>
    <tr>
      <td><strong>--fragment-len</strong></td>
      <td>Length of a fragment</td>
      <td>350</td>
      <td><code class="highlighter-rouge">--fragment-len 500</code></td>
    </tr>
    <tr>
      <td><strong>--perbase-errorrate</strong></td>
      <td>Sequencing error rate per base in percentage</td>
      <td>0.0 (%)</td>
      <td><code class="highlighter-rouge">--perbase-errorrate 0.2</code></td>
    </tr>
    <tr>
      <td><strong>--perbase-snprate</strong></td>
      <td>SNP rate per base in percentage</td>
      <td>0.0 (%)</td>
      <td><code class="highlighter-rouge">--perbase-snprate 0.1</code></td>
    </tr>
    <tr>
      <td><strong>--skip-fragment-regions</strong></td>
      <td>Comma separated list of regions from which no reads originate</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">--skip-fragment-regions 500-600,1200-1400</code></td>
    </tr>
    <tr>
      <td><strong>--random-seed</strong></td>
      <td>Seeding number for random number generation</td>
      <td>1</td>
      <td><code class="highlighter-rouge">--random-seed 10</code></td>
    </tr>
    <tr>
      <td><strong>--debug</strong></td>
      <td>Comma separated list of debugging parameters</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">--debug "pair,test_list:A*03:01:01:05,A*24:02:01:01"</code></td>
    </tr>
  </tbody>
</table>

<p>Examples:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="nv">$ </span>hisatgenotype_locus.py <span class="nt">--base</span> hla <span class="nt">--locus-list</span> A 
 <span class="nv">$ </span>hisatgenotype_locus.py <span class="nt">--base</span> hla <span class="nt">--locus-list</span> A <span class="nt">--debug</span> <span class="s2">"pair"</span>
 <span class="nv">$ </span>hisatgenotype_locus.py <span class="nt">--base</span> hla <span class="nt">--locus-list</span> B <span class="nt">--assembly</span> <span class="nt">--debug</span> <span class="s2">"pair,test_list:A*03:01:01:05,A*24:02:01:01"</span>
 <span class="nv">$ </span>hisatgenotype_locus.py <span class="nt">--base</span> hla <span class="nt">--locus-list</span> A <span class="nt">--assembly</span> <span class="nt">--simulate-interval</span> 10 <span class="nt">--perbase-errorrate</span> 0.3 <span class="nt">--perbase-snprate</span> 0.1 <span class="nt">--num-editdist</span> 2 <span class="nt">--debug</span> <span class="s2">"pair,full"</span>
 <span class="nv">$ </span>hisatgenotype_locus.py <span class="nt">--base</span> hla <span class="nt">--locus-list</span> A <span class="nt">--assembly</span> <span class="nt">--simulate-interval</span> 10 <span class="nt">--perbase-errorrate</span> 0.3 <span class="nt">--perbase-snprate</span> 0.1 <span class="nt">--num-editdist</span> 2 <span class="nt">--debug</span> <span class="s2">"pair,full"</span> <span class="nt">--skip-fragment-regions</span> 1000-1020 <span class="nt">--random-seed</span> 0
</code></pre></div></div>

<h2 id="hisatgenotype_build_genomepy--building-a-genotype-genome">hisatgenotype_build_genome.py – Building a genotype genome</h2>
<p>The hisatgenotype_build_genome.py python script will create a genotype genome using the human reference genome (GRCh38) and several genomic databases. A tutorial for the script can be found here.</p>

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Description</th>
      <th>Default</th>
      <th>Example</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>--base</strong></td>
      <td>Base file name for index, variants, haplotypes, etc.</td>
      <td>genotype_genome</td>
      <td><code class="highlighter-rouge">--base genotype_genome</code></td>
    </tr>
    <tr>
      <td><strong>--database-list</strong></td>
      <td>Comma separated list of databases</td>
      <td>hla</td>
      <td><code class="highlighter-rouge">--database-list hla,codis,cyp</code></td>
    </tr>
    <tr>
      <td><strong>--commonvar</strong></td>
      <td>Include common variants from dbSNP (about 13 million)</td>
      <td>false</td>
      <td><code class="highlighter-rouge">--commonvar</code></td>
    </tr>
    <tr>
      <td><strong>--clinvar</strong></td>
      <td>Include clinically relevant variants from NCBI ClinVar database</td>
      <td>false</td>
      <td><code class="highlighter-rouge">--clinvar</code></td>
    </tr>
    <tr>
      <td><strong>-p / --threads</strong></td>
      <td>Number of threads to be used</td>
      <td>1</td>
      <td><code class="highlighter-rouge">-p 4</code></td>
    </tr>
    <tr>
      <td><strong>--verbose</strong></td>
      <td>Provide more information</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--verbose</code></td>
    </tr>
    <tr>
      <td><strong>-h / --help</strong></td>
      <td>Output help message</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--help</code></td>
    </tr>
  </tbody>
</table>

<p>Examples:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="nv">$ </span>hisatgenotype_build_genome.py <span class="nt">-p</span> 4 <span class="nt">--base</span> genotype_genome <span class="nt">--database-list</span> hla <span class="nt">--commonvar</span>
 <span class="nv">$ </span>hisatgenotype_build_genome.py <span class="nt">-p</span> 4 <span class="nt">--base</span> genotype_genome <span class="nt">--database-list</span> hla,codis,cyp <span class="nt">--commonvar</span>
</code></pre></div></div>

<h2 id="hisatgenotype_extract_varspy--extraction-of-variants-haplotypes-etc">hisatgenotype_extract_vars.py – Extraction of variants, haplotypes, etc.</h2>
<p>This hisatgenotype_extract_vars.py python script is typically run by other scripts, though instructions are provided here for users who wish to directly run the script. The script will extract allele sequences, variants, haplotypes, etc. from multiple sequence alignments available in hisatgenotype_db. Tutorials for the script can be found at HLA typing and DNA fingerprinting anlaysis.</p>

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Description</th>
      <th>Default</th>
      <th>Example</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>--base</strong></td>
      <td>Base file name for index, variants, haplotypes, etc.</td>
      <td>hla</td>
      <td><code class="highlighter-rouge">--base codis</code></td>
    </tr>
    <tr>
      <td><strong>--locus-list</strong></td>
      <td>Comma separated list of loci</td>
      <td>(empty) meaning every locus available</td>
      <td><code class="highlighter-rouge">--locus-list A,B,C,DQA1,DQB1,DRB1</code></td>
    </tr>
    <tr>
      <td><strong>--min-var-freq</strong></td>
      <td>Exclude variants whose freq is below than this value in percentage</td>
      <td>0.0 (%)</td>
      <td><code class="highlighter-rouge">--min-var-freq 0.1</code></td>
    </tr>
    <tr>
      <td><strong>--leftshift</strong></td>
      <td>Shift deletions to the leftmost</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--leftmost</code></td>
    </tr>
    <tr>
      <td><strong>--inter-gap</strong></td>
      <td>Maximum distance for variants to be in the same haplotype</td>
      <td>30</td>
      <td><code class="highlighter-rouge">--inter-gap 30</code></td>
    </tr>
    <tr>
      <td><strong>--intra-gap</strong></td>
      <td>Break a haplotype into several haplotypes</td>
      <td>50</td>
      <td><code class="highlighter-rouge">--intra-gap 50</code></td>
    </tr>
    <tr>
      <td><strong>--verbose</strong></td>
      <td>Provide more information</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--verbose</code></td>
    </tr>
    <tr>
      <td><strong>-h / --help</strong></td>
      <td>Output help message</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--help</code></td>
    </tr>
  </tbody>
</table>

<p>Examples:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="nv">$ </span>hisatgenotype_extract_vars.py <span class="nt">--base</span> hla <span class="nt">--locus-list</span> A,B,C,DQA1,DQB1,DRB1 <span class="nt">--min-var-freq</span> 0.1
 <span class="nv">$ </span>hisatgenotype_extract_vars.py <span class="nt">--base</span> codis
</code></pre></div></div>

<h2 id="hisatgenotype_extract_readspy--extraction-of-reads-that-belong-to-genomic-regions-of-interest">hisatgenotype_extract_reads.py – Extraction of reads that belong to genomic regions of interest</h2>
<p>The hisatgenotype_extract_reads.py python script will extract reads that belong to loci of interest from whole sequencing reads. The script can also be used to extract reads from many samples. A tutorial for the script can be found here.</p>

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Description</th>
      <th>Default</th>
      <th>Example</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>--base</strong></td>
      <td>Base file name for index, variants, haplotypes, etc.</td>
      <td>genotype_genome</td>
      <td><code class="highlighter-rouge">--base genotype_genome</code></td>
    </tr>
    <tr>
      <td><strong>--database-list</strong></td>
      <td>Comma separated list of databases</td>
      <td>(empty) meaning every database available</td>
      <td><code class="highlighter-rouge">--database-list hla,codis,cyp</code></td>
    </tr>
    <tr>
      <td><strong>-U</strong></td>
      <td>Single-end read file name</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-U read.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-1</strong></td>
      <td>Paired-end read file name 1</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-1 read.1.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-2</strong></td>
      <td>Paired-end read file name 2</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">-2 read.2.fq.gz</code></td>
    </tr>
    <tr>
      <td><strong>--read-dir</strong></td>
      <td>Name of directory where read files are placed</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">--read-dir test_input</code></td>
    </tr>
    <tr>
      <td><strong>--out-dir</strong></td>
      <td>Name of directory where extracted read files will be placed</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">--out-dir test_output</code></td>
    </tr>
    <tr>
      <td><strong>--suffix</strong></td>
      <td>Read file suffix name</td>
      <td>fq.gz</td>
      <td><code class="highlighter-rouge">--suffix fastq.gz</code></td>
    </tr>
    <tr>
      <td><strong>-f / --fasta</strong></td>
      <td>Reads are provided in FASTA format</td>
      <td>false</td>
      <td><code class="highlighter-rouge">-f</code></td>
    </tr>
    <tr>
      <td><strong>--num-editdist</strong></td>
      <td>Number of maximum edit distance or mismatches allowed in the alignment, typing, and assembly of a read (not a pair)</td>
      <td>2</td>
      <td><code class="highlighter-rouge">--num-editdist 0</code></td>
    </tr>
    <tr>
      <td><strong>-p / --threads</strong></td>
      <td>Number of threads to be used</td>
      <td>1</td>
      <td><code class="highlighter-rouge">-p 4</code></td>
    </tr>
    <tr>
      <td><strong>--max-sample</strong></td>
      <td>Number of samples to be extracted</td>
      <td>sys.maxint</td>
      <td><code class="highlighter-rouge">--max-sample 917</code></td>
    </tr>
    <tr>
      <td><strong>--job-range</strong></td>
      <td>Comma separated two numbers representing a range</td>
      <td>0,1</td>
      <td><code class="highlighter-rouge">--job-range 1,4</code></td>
    </tr>
    <tr>
      <td><strong>--verbose</strong></td>
      <td>Provide more information</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--verbose</code></td>
    </tr>
    <tr>
      <td><strong>-h / --help</strong></td>
      <td>Output help message</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--help</code></td>
    </tr>
  </tbody>
</table>

<p>Examples:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="nv">$ </span>hisatgenotype_extract_reads.py <span class="nt">--base</span> genotype_genome <span class="nt">--database-list</span> hla,codis,cyp <span class="nt">-1</span> test.1.fq.gz <span class="nt">-2</span> test.2.fq.gz
 <span class="nv">$ </span>hisatgenotype_extract_reads.py <span class="nt">--base</span> genotype_genome <span class="nt">--database-list</span> hla,codis,cyp <span class="nt">--read-dir</span> test_input <span class="nt">--out-dir</span> test_output
</code></pre></div></div>

<h2 id="hisatgenotype_locus_samplespy--analysis-of-many-samples">hisatgenotype_locus_samples.py – Analysis of many samples</h2>
<p>The hisatgenotype_locus_samples.py python script will analyze genomic regions of interest using reads that are extracted from using hisatgenotype_extract_reads.py. Please avoid directly running the script on whole genomic sequencing reads as doing so will lead to inaccurate results due to reasons partly described here. A tutorial for the script can be found at HLA typing. The script is designed to analyze many samples.</p>

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Description</th>
      <th>Default</th>
      <th>Example</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>--base</strong></td>
      <td>Base file name for index, variants, haplotypes, etc.</td>
      <td>genotype_genome</td>
      <td><code class="highlighter-rouge">--base genotype_genome</code></td>
    </tr>
    <tr>
      <td><strong>--region-list</strong></td>
      <td>Comma separated list of regions</td>
      <td>(empty) meaning every region available</td>
      <td><code class="highlighter-rouge">--region-list hla.A,hla.B,hla.C,codis.FGA,cyp.2D6</code></td>
    </tr>
    <tr>
      <td><strong>--read-dir</strong></td>
      <td>Name of directory where read files are placed</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">--read-dir test_input</code></td>
    </tr>
    <tr>
      <td><strong>--out-dir</strong></td>
      <td>Output directory name</td>
      <td><em>None</em></td>
      <td><code class="highlighter-rouge">--out-dir test_output</code></td>
    </tr>
    <tr>
      <td><strong>-f / --fasta</strong></td>
      <td>Reads are provided in FASTA format</td>
      <td>false</td>
      <td><code class="highlighter-rouge">-f</code></td>
    </tr>
    <tr>
      <td><strong>--num-editdist</strong></td>
      <td>Number of maximum edit distance or mismatches allowed in the alignment, typing, and assembly of a read (not a pair)</td>
      <td>2</td>
      <td><code class="highlighter-rouge">--num-editdist 0</code></td>
    </tr>
    <tr>
      <td><strong>-p / --threads</strong></td>
      <td>Number of threads to be used</td>
      <td>1</td>
      <td><code class="highlighter-rouge">-p 4</code></td>
    </tr>
    <tr>
      <td><strong>--assembly</strong></td>
      <td>Perform assembly of each locus of interest</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--assembly</code></td>
    </tr>
    <tr>
      <td><strong>--verbose</strong></td>
      <td>Provide more information</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--verbose</code></td>
    </tr>
    <tr>
      <td><strong>-h / --help</strong></td>
      <td>Output help message</td>
      <td>disabled</td>
      <td><code class="highlighter-rouge">--help</code></td>
    </tr>
  </tbody>
</table>

<p>Examples:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code> <span class="nv">$ </span>hisatgenotype_locus_samples.py <span class="nt">-p</span> 4 <span class="nt">--region-list</span> hla <span class="nt">--read-dir</span> test_input <span class="nt">--out-dir</span> test_output
 <span class="nv">$ </span>hisatgenotype_locus_samples.py <span class="nt">-p</span> 4 <span class="nt">--region-list</span> hla.A,hla.B <span class="nt">--read-dir</span> test_input <span class="nt">--out-dir</span> test_output
 <span class="nv">$ </span>hisatgenotype_locus_samples.py <span class="nt">-p</span> 4 <span class="nt">--region-list</span> codis <span class="nt">--read-dir</span> test_input <span class="nt">--out-dir</span> test_output
</code></pre></div></div>

        </section>
        <footer>
    
</footer>

    </article>
</div>

    </div>

    <aside class="site-aside">
        <div class="inner">
            <div class="block">
                <form action="/hisat-genotype/search">
                    <input type="search" id="search" name="q" placeholder="Search" />
                </form>
            </div>
            <ul>
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                <li><a class="page-link" href="/hisat-genotype/main/">Main</a></li>
                
                
                
                <li><a class="page-link" href="/hisat-genotype/about/">About</a></li>
                
                
                
                
                
                
                
                <li><a class="page-link" href="/hisat-genotype/manual/">Manual</a></li>
                
                
                
                
                
                <li><a class="page-link" href="/hisat-genotype/tutorials/">Tutorials</a></li>
                
                
                
                
            </ul>
            <ul class="icons">
                


<li>
    <a href="/hisat-genotype/feed.xml">
        <i class="fa fa-fw fa-rss"></i>
    </a>
</li>





<li>
    <a href="https://github.com/ttskch">
        <i class="fa fa-fw fa-github"></i>
    </a>
</li>





<li>
    <a href="https://twitter.com/ttskch">
        <i class="fa fa-fw fa-twitter"></i>
    </a>
</li>
































            </ul>
            <hr class="with-no-margin margin-bottom"/>

            <div class="block sticky">
                <h2>Recent posts</h2>
                <ul>
                    
                    
                    
                    
                    <li><a href="/hisat-genotype/2015/01/01/hello/">Hello, Jekyll and TtskchTheme</a></li>
                    
                </ul>
            </div>
        </div>
    </aside>
</div>

<footer class="site-footer">
    <div class="inner">
        <span>Powered by <a href="http://jekyllrb.com">Jekyll</a> with <a href="https://github.com/ttskch/jekyll-ttskch-theme">TtskchTheme</a></span>
    </div>
</footer>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="/hisat-genotype/assets/lib/garand-sticky/jquery.sticky.js"></script>
<script src="/hisat-genotype/assets/js/script.js"></script>





<!-- Init Facebook SDK -->


<!-- Google Analytics -->



</body>
</html>
