<!DOCTYPE html>
<html lang="en">
<head>
    

    




    

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <title>Manual | HISAT-genotype</title>

    <meta name="description" content="next-generation genomic analysis software platform">

    <link rel="shortcut icon" href="/hisat-genotype/assets/img/ogp.png" type="image/x-icon">
    <link rel="canonical" href="http://localhost:4000/hisat-genotype/manual/">
    <link rel="alternate" type="application/atom+xml" title="HISAT-genotype" href="/hisat-genotype/feed.xml" />

    

    <meta property="og:title" content="Manual | HISAT-genotype" />
    <meta property="og:type" content="website" />
    <meta property="og:image" content="//ttskch.github.io/jekyll-ttskch-theme/assets/img/ogp.png" />
    <meta property="og:url" content="http://localhost:4000/hisat-genotype/manual/" />
    <meta property="og:site_name" content="HISAT-genotype" />
    <meta property="fb:admins" content="" />
    <meta property="fb:app_id" content="" />
    <meta property="og:description" content="next-generation genomic analysis software platform" />

    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
    <script src="https://use.fontawesome.com/1f5f360d80.js"></script>
    <link href="//fonts.googleapis.com/css?family=Source+Sans+Pro:400,700,700italic,400italic" rel="stylesheet">

    <link href="/hisat-genotype/assets/css/style.css" rel="stylesheet">
</head>
<body>

<header class="site-header">
    <div class="inner clearfix">
        
        <a href="/hisat-genotype/">
            <img class="avatar" src="/hisat-genotype/assets/img/ogp.png" alt=""/>
        </a>
        
        <h1 class="clearfix">
            <a class="title " href="/hisat-genotype/">HISAT-genotype</a>
            <br><span class="description">next-generation genomic analysis software platform</span>
        </h1>
    </div>
</header>

<div class="site-container">
    <div class="site-content">
        <div class="article-wrapper">
    <article>
        

<header>
    <div class="panel">
        <h1>
            
            Manual
            
        </h1>

        <ul class="tags">
            
            
            
        </ul>

        <div class="clearfix">
            <ul class="meta">
                

                
            </ul>
        </div>
    </div>

    

    
</header>

        <section class="post-content">
            
<ul id="markdown-toc">
  <li><a href="#introduction" id="markdown-toc-introduction">Introduction</a></li>
  <li><a href="#hisatgenotype-set-up" id="markdown-toc-hisatgenotype-set-up">HISATgenotype Set-up</a>    <ul>
      <li><a href="#requirements" id="markdown-toc-requirements">Requirements</a></li>
      <li><a href="#downloading-hisatgenotype-and-building-hisat2-from-source" id="markdown-toc-downloading-hisatgenotype-and-building-hisat2-from-source">Downloading HISATgenotype and Building HISAT2 from source</a></li>
      <li><a href="#adding-hisatgenotype-to-path" id="markdown-toc-adding-hisatgenotype-to-path">Adding HISATgenotype to PATH</a></li>
    </ul>
  </li>
  <li><a href="#running-hisatgenotype" id="markdown-toc-running-hisatgenotype">Running HISATgenotype</a>    <ul>
      <li><a href="#past-manuals" id="markdown-toc-past-manuals">Past Manuals</a></li>
      <li><a href="#hisatgenotype---analysis-of-a-whole-human-genome" id="markdown-toc-hisatgenotype---analysis-of-a-whole-human-genome">hisatgenotype - Analysis of a whole human genome</a>        <ul>
          <li><a href="#standard-options" id="markdown-toc-standard-options">Standard Options</a></li>
          <li><a href="#advanced-options" id="markdown-toc-advanced-options">Advanced Options</a></li>
        </ul>
      </li>
      <li><a href="#hisatgenotype_toolkit---individual-scripts-and-tools-for-custom-pipelines" id="markdown-toc-hisatgenotype_toolkit---individual-scripts-and-tools-for-custom-pipelines">hisatgenotype_toolkit - Individual Scripts and Tools for Custom Pipelines</a></li>
    </ul>
  </li>
</ul>

<h1 id="introduction">Introduction</h1>

<p>Using the human reference genome, referred to as the linear reference (e.g. GRCh38 and hg38), for genomic analysis would be rather straightforward if our variants were uniformly distributed with only one nucleotide difference every 1,000 nucleotides, which most of the currently used alignment programs could handle with great accuracy. However, the distribution of variants is not uniform. Some genomic regions such as HLA genes and DNA fingerprinting loci are highly polymorphic. So using the reference genome for analyzing such highly polymorphic regions may not be the most effective approach, and this is where our graph reference comes into play.</p>

<hr />

<h1 id="hisatgenotype-set-up">HISATgenotype Set-up</h1>

<p>We use <a href="http://ccb.jhu.edu/software/hisat2">HISAT2</a> for graph representation and alignment, which is currently the most practical and quickest program available. We refer to hisat-genotype as our top directory where all of our programs are located. hisat-genotype is a place holder that you can change to whatever name youâ€™d like to use.</p>

<h2 id="requirements">Requirements</h2>

<p>HISATgenotype is encoded in python2 and contains standard python libraries. A
python 2.7 release is recommended. Additional software required is <a href="https://www.htslib.org">Samtools</a>
version 1.3 or later.</p>

<p>Building HISAT2 from source requires a GNU-like environment with GCC, GNU Make
and other basics.  It should be possible to build HISAT2 on most vanilla Linux
installations or on a Mac installation with <a href="https://developer.apple.com/Xcode">Xcode</a> installed.  HISAT2 can
also be built on Windows using <a href="http://www.cygwin.com/">Cygwin</a> or <a href="http://www.mingw.org/">MinGW</a> (MinGW recommended). For a
MinGW build the choice of what compiler is to be used is important since this
will determine if a 32 or 64 bit code can be successfully compiled using it. If
there is a need to generate both 32 and 64 bit on the same machine then a multilib
MinGW has to be properly installed. <a href="http://www.mingw.org/wiki/msys">MSYS</a>, the <a href="http://cygwin.com/packages/mingw-zlib/">zlib</a> library, and depending on
architecture <a href="http://sourceware.org/pthreads-win32/">pthreads</a> library are also required. We are recommending a 64 bit
build since it has some clear advantages in real life research problems. In order
to simplify the MinGW setup it might be worth investigating popular MinGW personal
builds since these are coming already prepared with most of the toolchains needed.</p>

<h2 id="downloading-hisatgenotype-and-building-hisat2-from-source">Downloading HISATgenotype and Building HISAT2 from source</h2>
<p>This download example will place HISATgenotype in your home (~) directory if you are using a linux system. 
Change the ~ to whichever directory you desire if this is not the behavior you want.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>git clone https://github.com/DaehwanKimLab/hisat-genotype ~/hisat-genotype
<span class="nv">$ </span><span class="nb">cd</span> ~/hisat-genotype
hisat-genotype<span class="nv">$ </span>git checkout hisatgenotype_v1.1.2_beta
<span class="nv">$ </span>make hisat2-align-s hisat2-build-s hisat2-inspect-s
</code></pre></div></div>

<h2 id="adding-hisatgenotype-to-path">Adding HISATgenotype to PATH</h2>
<p>Add the above directory (hisat-genotype) to your PATH environment variable
(e.g. ~/.bashrc) to make the binaries built above and other python scripts
available everywhere:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">export </span><span class="nv">PATH</span><span class="o">=</span>~/hisat-genotype:~/hisat-genotype/hisatgenotype_scripts:<span class="nv">$PATH</span>
<span class="nv">$ </span><span class="nb">export </span><span class="nv">PYTHONPATH</span><span class="o">=</span>~/hisat-genotype/hisatgenotype_modules:<span class="nv">$PYTHONPATH</span>
</code></pre></div></div>

<hr />

<h1 id="running-hisatgenotype">Running HISATgenotype</h1>

<h2 id="past-manuals">Past Manuals</h2>

<p><a href="/hisat-genotype/manual_old/">Manual Pre v1.1</a></p>

<h2 id="hisatgenotype---analysis-of-a-whole-human-genome">hisatgenotype - Analysis of a whole human genome</h2>
<p>The hisatgenotype.py python script will analyze a whole human genome using whole genome sequencing reads. It will align reads to genotype genome, extract reads belonging to each locus of interest, and perform typing and assembly.</p>

<h3 id="standard-options">Standard Options</h3>

<ul>
  <li><strong>-x / --ref-genome</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>Base name for genome index if not genotype_genome. Generally reserved for custom graph genomes end user may want to use<br />
Example: <code class="highlighter-rouge">-x custome_genome</code></p>
    </blockquote>
  </li>
  <li><strong>--base / --base-fname</strong> | <em>Default</em> : (empty) all databases
    <blockquote>
      <p>Base file name for index, variants, haplotypes, etc. (e.g. hla, rbg, codis). This will be anything in the hisatgenotype_db folder<br />
Example: <code class="highlighter-rouge">--base hla</code></p>
    </blockquote>
  </li>
  <li><strong>--locus-list</strong> | <em>Default</em> : (empty) all genes
    <blockquote>
      <p>A comma-separated list of gene names
Example: <code class="highlighter-rouge">--locus-list A,B,C,DRB1,DQA1,DQB1</code></p>
    </blockquote>
  </li>
  <li><strong>-f / --fasta</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Bool to indicate if reads are provided in FASTA format<br />
Example: <code class="highlighter-rouge">-f</code></p>
    </blockquote>
  </li>
  <li><strong>-U</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>Single-end read file name<br />
Example: <code class="highlighter-rouge">-U read.fq.gz</code></p>
    </blockquote>
  </li>
  <li><strong>-1</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>Paired-end read file name 1<br />
Example: <code class="highlighter-rouge">-1 read.1.fq.gz</code></p>
    </blockquote>
  </li>
  <li><strong>-2</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>Paired-end read file name 2<br />
Example: <code class="highlighter-rouge">-2 read.2.fq.gz</code></p>
    </blockquote>
  </li>
  <li><strong>--keep-alignemnt</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Bool to keep the alignment BAM file if typing from FASTQ(A). If typing from a BAM file this is irrelevent.<br />
Example: <code class="highlighter-rouge">--keep-alignment</code></p>
    </blockquote>
  </li>
  <li><strong>--in-dir</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>Directory HISATgenotpye will search for FASTQ(A) files and batch process. HISATgenotype will attempt to automatically pair the files if <code class="highlighter-rouge">--single-end</code> isnâ€™t set.
Try to have the names be similar between the pairs with a single difference to make it easier for HISATgenotype to pair the files (e.g. <em>hg_granulocyte_samp1_L.fastq</em> and <em>hg_granulocyte_samp1_R.fastq</em>)<br />
Example: <code class="highlighter-rouge">--in-dir input_fastq_dir</code></p>
    </blockquote>
  </li>
  <li><strong>--out-dir</strong> | <em>Default</em> : <code class="highlighter-rouge">/hisatgenotype_out</code>
    <blockquote>
      <p>Directory where all resulting files will be placed. This includes the typing results, BAM files, and assembly files.<br />
Example: <code class="highlighter-rouge">--out-dir results</code></p>
    </blockquote>
  </li>
  <li><strong>--bamfile</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>BAM file name if using already aligned reads<br />
Example: <code class="highlighter-rouge">--bamfile hg_granulocyte_samp1.bam</code></p>
    </blockquote>
  </li>
  <li><strong>--single-end</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Bool to indicate if file(s) is/are single ended. Only needed with <code class="highlighter-rouge">--in-dir</code> or <code class="highlighter-rouge">--bamfile</code><br />
Example: <code class="highlighter-rouge">--single-end</code></p>
    </blockquote>
  </li>
  <li><strong>--assembly</strong> | <em>Default</em> : disabled
    <blockquote>
      <p>Perform assembly of each locus of interest<br />
Example: <code class="highlighter-rouge">--assembly</code></p>
    </blockquote>
  </li>
  <li><strong>--assembly-name</strong> | <em>Default</em> : <code class="highlighter-rouge">assemply_graph</code>
    <blockquote>
      <p>Assembly base file name to use<br />
Example: <code class="highlighter-rouge">--assembly-name hg_granulocyte</code></p>
    </blockquote>
  </li>
  <li><strong>--assembly-verbose</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Bool to output additional assembly information<br />
Example: <code class="highlighter-rouge">--assembly-verbose</code></p>
    </blockquote>
  </li>
  <li><strong>-p / --threads</strong> | <em>Default</em> : 1
    <blockquote>
      <p>Number of threads to be used<br />
Example: <code class="highlighter-rouge">-p 4</code></p>
    </blockquote>
  </li>
  <li><strong>--verbose</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Provide more information
Example: <code class="highlighter-rouge">--verbose</code></p>
    </blockquote>
  </li>
  <li><strong>-h / --help</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Output help message<br />
Example: <code class="highlighter-rouge">--help </code></p>
    </blockquote>
  </li>
  <li><strong>--advanced-help</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Output help message for advanced options<br />
Example: <code class="highlighter-rouge">--advanced-help</code></p>
    </blockquote>
  </li>
</ul>

<h3 id="advanced-options">Advanced Options</h3>

<ul>
  <li><strong>--keep-extract</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Bool to keep extracted read fastq files<br />
Example: <code class="highlighter-rouge">--keep-extract</code></p>
    </blockquote>
  </li>
  <li><strong>--build-base</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Build the indexes listed in --base-fname</p>
    </blockquote>
  </li>
  <li><strong>--aligner</strong> | <em>Default</em> : <code class="highlighter-rouge">hisat2</code>
    <blockquote>
      <p>Set aligner to use (ex. hisat2, bowtie2)</p>
    </blockquote>
  </li>
  <li><strong>--linear-index</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Use linear index</p>
    </blockquote>
  </li>
  <li><strong>--num-mismatch</strong> | <em>Default</em> : <code class="highlighter-rouge">0</code>
    <blockquote>
      <p>Maximum number of mismatches per read alignment to be considered</p>
    </blockquote>
  </li>
  <li><strong>--inter-gap</strong>
    <blockquote>
      <p>Maximum distance for variants to be in the same haplotype</p>
    </blockquote>
  </li>
  <li><strong>--intra-gap</strong>
    <blockquote>
      <p>Break a haplotype into several haplotypes</p>
    </blockquote>
  </li>
  <li><strong>--whole-haplotype</strong>
    <blockquote>
      <p>Include partial alleles (e.g. A_nuc.fasta)</p>
    </blockquote>
  </li>
  <li><strong>--min-var-freq</strong> | <em>Default</em> : <code class="highlighter-rouge">0.0</code>
    <blockquote>
      <p>Exclude variants whose freq is below than this value in percentage</p>
    </blockquote>
  </li>
  <li><strong>--ext-seq</strong> | <em>Default</em> : <code class="highlighter-rouge">0</code>
    <blockquote>
      <p>Length of extra sequences flanking backbone sequences</p>
    </blockquote>
  </li>
  <li><strong>--leftshift</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Shift deletions to the leftmost</p>
    </blockquote>
  </li>
  <li><strong>--suffix</strong> | <em>Default</em> : <code class="highlighter-rouge">fq.gz</code>
    <blockquote>
      <p>Read file suffix</p>
    </blockquote>
  </li>
  <li><strong>--simulation</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Simulated reads (Default: False)</p>
    </blockquote>
  </li>
  <li><strong>--pp, --threads-aprocess</strong> | <em>Default</em> : <code class="highlighter-rouge">1</code>
    <blockquote>
      <p>Number of threads a process</p>
    </blockquote>
  </li>
  <li><strong>--max-sample</strong> | <em>Default</em> : <code class="highlighter-rouge">sys.maxint</code>
    <blockquote>
      <p>Number of samples to be extracted</p>
    </blockquote>
  </li>
  <li><strong>--job-range</strong> | <em>Default</em> : <code class="highlighter-rouge">0,1</code>
    <blockquote>
      <p>two numbers (e.g. 1,3)</p>
    </blockquote>
  </li>
  <li><strong>--extract-whole</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Extract all reads</p>
    </blockquote>
  </li>
  <li><strong>--no-partial</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Include partial alleles (e.g. A_nuc.fasta)</p>
    </blockquote>
  </li>
  <li><strong>--simulate-interval</strong> | <em>Default</em> : <code class="highlighter-rouge">10</code>
    <blockquote>
      <p>Reads simulated at every these base pairs</p>
    </blockquote>
  </li>
  <li><strong>--read-len</strong> | <em>Default</em> : <code class="highlighter-rouge">100</code>
    <blockquote>
      <p>Length of simulated reads</p>
    </blockquote>
  </li>
  <li><strong>--fragment-len</strong> | <em>Default</em> : `350
    <blockquote>
      <p>Length of fragments</p>
    </blockquote>
  </li>
  <li><strong>--best-alleles</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p><em>Placeholder</em></p>
    </blockquote>
  </li>
  <li><strong>--random-seed</strong> | <em>Default</em> : <code class="highlighter-rouge">1</code>
    <blockquote>
      <p>A seeding number for randomness</p>
    </blockquote>
  </li>
  <li><strong>--num-editdist</strong> | <em>Default</em> : <code class="highlighter-rouge">2</code>
    <blockquote>
      <p>Maximum number of mismatches per read alignment to be considered</p>
    </blockquote>
  </li>
  <li><strong>--perbase-errorrate</strong> | <em>Default</em> : <code class="highlighter-rouge">0.0</code>
    <blockquote>
      <p>Per basepair error rate in percentage when simulating</p>
    </blockquote>
  </li>
  <li><strong>--perbase-snprate</strong> | <em>Default</em> : <code class="highlighter-rouge">0.0</code>
    <blockquote>
      <p>Per basepair SNP rate in percentage when simulating</p>
    </blockquote>
  </li>
  <li><strong>--skip-fragment-regions</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>A comma-separated list of regions from which no reads originate, e.g., 500-600,1200-1400</p>
    </blockquote>
  </li>
  <li><strong>--verbose-level</strong> | <em>Default</em> : <code class="highlighter-rouge">0</code>
    <blockquote>
      <p>also print some statistics to stderr</p>
    </blockquote>
  </li>
  <li><strong>--no-error-correction</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Correct sequencing errors</p>
    </blockquote>
  </li>
  <li><strong>--only-locus-list</strong> | <em>Default</em> : (empty) all genes
    <blockquote>
      <p>A comma-separated list of genes</p>
    </blockquote>
  </li>
  <li><strong>--discordant</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Allow discordantly mapped pairs or singletons</p>
    </blockquote>
  </li>
  <li><strong>--type-primary-exons</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Look at primary exons first</p>
    </blockquote>
  </li>
  <li><strong>--keep-low-abundance-alleles</strong> | <em>Default</em> : <code class="highlighter-rouge">False</code>
    <blockquote>
      <p>Do not remove alleles with low abundance while performing typing</p>
    </blockquote>
  </li>
  <li><strong>--display-alleles</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>A comma-separated list of alleles to display in HTML</p>
    </blockquote>
  </li>
  <li><strong>--debug</strong> | <em>Default</em> : <em>None</em>
    <blockquote>
      <p>Test database or code<br />
(options: basic, pair, full, single-end, test_list, test_id)<br />
Example: <code class="highlighter-rouge">--debug test_id:10,basic</code></p>
    </blockquote>
  </li>
</ul>

<h2 id="hisatgenotype_toolkit---individual-scripts-and-tools-for-custom-pipelines">hisatgenotype_toolkit - Individual Scripts and Tools for Custom Pipelines</h2>

<p><em>Work In Progress to document</em></p>


        </section>
        <footer>
    
</footer>

    </article>
</div>

    </div>

    <aside class="site-aside">
        <div class="inner">
            <div class="block">
                <form action="/hisat-genotype/search">
                    <input type="search" id="search" name="q" placeholder="Search" />
                </form>
            </div>
            <ul>
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                <li><a class="page-link" href="/hisat-genotype/main/">Main</a></li>
                
                
                
                <li><a class="page-link" href="/hisat-genotype/about/">About</a></li>
                
                
                
                
                
                
                
                <li><a class="page-link" href="/hisat-genotype/manual/">Manual</a></li>
                
                
                
                
                
                <li><a class="page-link" href="/hisat-genotype/tutorials/">Tutorials</a></li>
                
                
                
                
            </ul>
            <ul class="icons">
                


<li>
    <a href="/hisat-genotype/feed.xml">
        <i class="fa fa-fw fa-rss"></i>
    </a>
</li>





<li>
    <a href="https://github.com/ttskch">
        <i class="fa fa-fw fa-github"></i>
    </a>
</li>





<li>
    <a href="https://twitter.com/ttskch">
        <i class="fa fa-fw fa-twitter"></i>
    </a>
</li>
































            </ul>
            <hr class="with-no-margin margin-bottom"/>

            <div class="block sticky">
                <h2>Recent posts</h2>
                <ul>
                    
                    
                    
                    
                    <li><a href="/hisat-genotype/2015/01/01/hello/">Hello, Jekyll and TtskchTheme</a></li>
                    
                </ul>
            </div>
        </div>
    </aside>
</div>

<footer class="site-footer">
    <div class="inner">
        <span>Powered by <a href="http://jekyllrb.com">Jekyll</a> with <a href="https://github.com/ttskch/jekyll-ttskch-theme">TtskchTheme</a></span>
    </div>
</footer>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="/hisat-genotype/assets/lib/garand-sticky/jquery.sticky.js"></script>
<script src="/hisat-genotype/assets/js/script.js"></script>





<!-- Init Facebook SDK -->


<!-- Google Analytics -->



</body>
</html>
