<!DOCTYPE html>
<html lang="en">
<head>
    

    




    

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <title>Manual | HISAT-genotype</title>

    <meta name="description" content="next-generation genomic analysis software platform">

    <link rel="shortcut icon" href="/hisatgenotype/assets/img/ogp.png" type="image/x-icon">
    <link rel="canonical" href="http://localhost:4000/hisatgenotype/manual/">
    <link rel="alternate" type="application/atom+xml" title="HISAT-genotype" href="/hisatgenotype/feed.xml" />

    

    <meta property="og:title" content="Manual | HISAT-genotype" />
    <meta property="og:type" content="website" />
    <meta property="og:image" content="//ttskch.github.io/jekyll-ttskch-theme/assets/img/ogp.png" />
    <meta property="og:url" content="http://localhost:4000/hisatgenotype/manual/" />
    <meta property="og:site_name" content="HISAT-genotype" />
    <meta property="fb:admins" content="" />
    <meta property="fb:app_id" content="" />
    <meta property="og:description" content="next-generation genomic analysis software platform" />

    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
    <script src="https://use.fontawesome.com/1f5f360d80.js"></script>
    <link href="//fonts.googleapis.com/css?family=Source+Sans+Pro:400,700,700italic,400italic" rel="stylesheet">

    <link href="/hisatgenotype/assets/css/style.css" rel="stylesheet">
</head>
<body>

<header class="site-header">
    <div class="inner clearfix">
        
        <a href="/hisatgenotype/">
            <img class="avatar" src="/hisatgenotype/assets/img/ogp.png" alt=""/>
        </a>
        
        <h1 class="clearfix">
            <a class="title " href="/hisatgenotype/">HISAT-genotype</a>
            <br><span class="description">next-generation genomic analysis software platform</span>
        </h1>
    </div>
</header>

<div class="site-container">
    <div class="site-content">
        <div class="article-wrapper">
    <article>
        

<header>
    <div class="panel">
        <h1>
            
            Manual
            
        </h1>

        <ul class="tags">
            
            
            
        </ul>

        <div class="clearfix">
            <ul class="meta">
                

                
            </ul>
        </div>
    </div>

    

    
</header>

        <section class="post-content">
            <h1 id="introduction">Introduction</h1>

<h2 id="what-is-hisat2">What is HISAT2?</h2>

<p>HISAT2 is a fast and sensitive alignment program for mapping next-generation sequencing reads 
(whole-genome, transcriptome, and exome sequencing data) against the general human population 
(as well as against a single reference genome). Based on <a href="http://ieeexplore.ieee.org/xpls/abs_all.jsp?arnumber=6698337&amp;tag=1">GCSA</a> (an extension of <a href="http://en.wikipedia.org/wiki/Burrows-Wheeler_transform">BWT</a> for a graph), we designed and implemented a graph FM index (GFM),
an original approach and its first implementation to the best of our knowledge. 
In addition to using one global GFM index that represents general population, 
HISAT2 uses a large set of small GFM indexes that collectively cover the whole genome 
(each index representing a genomic region of 56 Kbp, with 55,000 indexes needed to cover human population). 
These small indexes (called local indexes) combined with several alignment strategies enable effective alignment of sequencing reads. 
This new indexing scheme is called Hierarchical Graph FM index (HGFM). 
We have developed HISAT 2 based on the <a href="http://ccb.jhu.edu/software/hisat">HISAT</a> and <a href="http://bowtie-bio.sf.net/bowtie2">Bowtie2</a> implementations.
HISAT2 outputs alignments in <a href="http://samtools.sourceforge.net/SAM1.pdf">SAM</a> format, enabling interoperation with a large number of other tools (e.g. <a href="http://samtools.sourceforge.net">SAMtools</a>, <a href="http://www.broadinstitute.org/gsa/wiki/index.php/The_Genome_Analysis_Toolkit">GATK</a>) that use SAM.
HISAT2 is distributed under the <a href="http://www.gnu.org/licenses/gpl-3.0.html">GPLv3 license</a>, and it runs on the command line under
Linux, Mac OS X and Windows.</p>

<h1 id="obtaining-hisat2">Obtaining HISAT2</h1>

<p>Download HISAT2 sources and binaries from the Releases sections on the right side.
Binaries are available for Intel architectures (<code class="highlighter-rouge">x86_64</code>) running Linux, and Mac OS X.</p>

<h2 id="building-from-source">Building from source</h2>

<p>Building HISAT2 from source requires a GNU-like environment with GCC, GNU Make
and other basics.  It should be possible to build HISAT2 on most vanilla Linux
installations or on a Mac installation with <a href="http://developer.apple.com/xcode/">Xcode</a> installed.  HISAT2 can
also be built on Windows using <a href="http://www.cygwin.com/">Cygwin</a> or <a href="http://www.mingw.org/">MinGW</a> (MinGW recommended). For a 
MinGW build the choice of what compiler is to be used is important since this
will determine if a 32 or 64 bit code can be successfully compiled using it. If 
there is a need to generate both 32 and 64 bit on the same machine then a multilib 
MinGW has to be properly installed. <a href="http://www.mingw.org/wiki/msys">MSYS</a>, the <a href="http://cygwin.com/packages/mingw-zlib/">zlib</a> library, and depending on 
architecture <a href="http://sourceware.org/pthreads-win32/">pthreads</a> library are also required. We are recommending a 64 bit
build since it has some clear advantages in real life research problems. In order 
to simplify the MinGW setup it might be worth investigating popular MinGW personal 
builds since these are coming already prepared with most of the toolchains needed.</p>

<p>First, download the <a href="http://ccb.jhu.edu/software/hisat2/downloads/hisat2-2.0.0-beta-source.zip">source package</a> from the Releases section on the right side.
Unzip the file, change to the unzipped directory, and build the
HISAT2 tools by running GNU <code class="highlighter-rouge">make</code> (usually with the command <code class="highlighter-rouge">make</code>, but
sometimes with <code class="highlighter-rouge">gmake</code>) with no arguments.  If building with MinGW, run <code class="highlighter-rouge">make</code>
from the MSYS environment.</p>

<p>HISAT2 is using the multithreading software model in order to speed up 
execution times on SMP architectures where this is possible. On POSIX 
platforms (like linux, Mac OS, etc) it needs the pthread library. Although
it is possible to use pthread library on non-POSIX platform like Windows, due
to performance reasons HISAT2 will try to use Windows native multithreading
if possible.</p>

<p>For the support of SRA data access in HISAT2, please download and install the <a href="https://github.com/ncbi/ngs/wiki/Downloads">NCBI-NGS</a> toolkit.
When running <code class="highlighter-rouge">make</code>, specify additional variables as follow.
<code class="highlighter-rouge">make USE_SRA=1 NCBI_NGS_DIR=/path/to/NCBI-NGS-directory NCBI_VDB_DIR=/path/to/NCBI-NGS-directory</code>,
where <code class="highlighter-rouge">NCBI_NGS_DIR</code> and <code class="highlighter-rouge">NCBI_VDB_DIR</code> will be used in Makefile for -I and -L compilation options.
For example, $(NCBI_NGS_DIR)/include and $(NCBI_NGS_DIR)/lib64 will be used.</p>

<h1 id="running-hisat2">Running HISAT2</h1>

<h2 id="adding-to-path">Adding to PATH</h2>

<p>By adding your new HISAT2 directory to your <a href="http://en.wikipedia.org/wiki/PATH_(variable)">PATH environment variable</a>, you
ensure that whenever you run <code class="highlighter-rouge">hisat2</code>, <code class="highlighter-rouge">hisat2-build</code> or <code class="highlighter-rouge">hisat2-inspect</code>
from the command line, you will get the version you just installed without
having to specify the entire path.  This is recommended for most users.  To do
this, follow your operating system’s instructions for adding the directory to
your <a href="http://en.wikipedia.org/wiki/PATH_(variable)">PATH</a>.</p>

<p>If you would like to install HISAT2 by copying the HISAT2 executable files
to an existing directory in your <a href="http://en.wikipedia.org/wiki/PATH_(variable)">PATH</a>, make sure that you copy all the
executables, including <code class="highlighter-rouge">hisat2</code>, <code class="highlighter-rouge">hisat2-align-s</code>, <code class="highlighter-rouge">hisat2-align-l</code>, <code class="highlighter-rouge">hisat2-build</code>, <code class="highlighter-rouge">hisat2-build-s</code>, <code class="highlighter-rouge">hisat2-build-l</code>, <code class="highlighter-rouge">hisat2-inspect</code>, <code class="highlighter-rouge">hisat2-inspect-s</code> and
<code class="highlighter-rouge">hisat2-inspect-l</code>.</p>

<h2 id="reporting">Reporting</h2>

<p>The reporting mode governs how many alignments HISAT2 looks for, and how to
report them.</p>

<p>In general, when we say that a read has an alignment, we mean that it has a
<a href="#valid-alignments-meet-or-exceed-the-minimum-score-threshold">valid alignment</a>.  When we say that a read has multiple alignments, we mean
that it has multiple alignments that are valid and distinct from one another.</p>

<p>By default, HISAT2 may soft-clip reads near their 5’ and 3’ ends.  Users can control this behavior by setting different penalties for soft-clipping (<a href="#hisat2-options-no-softclip"><code class="highlighter-rouge">--sp</code></a>) or by disallowing soft-clipping ([<code class="highlighter-rouge">--no-softclip</code>]).</p>

<h3 id="distinct-alignments-map-a-read-to-different-places">Distinct alignments map a read to different places</h3>

<p>Two alignments for the same individual read are “distinct” if they map the same
read to different places.  Specifically, we say that two alignments are distinct
if there are no alignment positions where a particular read offset is aligned
opposite a particular reference offset in both alignments with the same
orientation.  E.g. if the first alignment is in the forward orientation and
aligns the read character at read offset 10 to the reference character at
chromosome 3, offset 3,445,245, and the second alignment is also in the forward
orientation and also aligns the read character at read offset 10 to the
reference character at chromosome 3, offset 3,445,245, they are not distinct
alignments.</p>

<p>Two alignments for the same pair are distinct if either the mate 1s in the two
paired-end alignments are distinct or the mate 2s in the two alignments are
distinct or both.</p>

<h3 id="default-mode-search-for-one-or-more-alignments-report-each">Default mode: search for one or more alignments, report each</h3>

<p>HISAT2 searches for up to N distinct, primary alignments for
each read, where N equals the integer specified with the <code class="highlighter-rouge">-k</code> parameter.
Primary alignments mean alignments whose alignment score is equal or higher than any other alignments.
It is possible that multiple distinct alignments have the same score.
That is, if <code class="highlighter-rouge">-k 2</code> is specified, HISAT2 will search for at most 2 distinct
alignments. The alignment score for a paired-end alignment equals the sum of the
alignment scores of the individual mates.  Each reported read or pair alignment
beyond the first has the SAM ‘secondary’ bit (which equals 256) set in its FLAGS
field.  See the <a href="http://samtools.sourceforge.net/SAM1.pdf">SAM specification</a> for details.</p>

<p>HISAT2 does not “find” alignments in any specific order, so for reads that
have more than N distinct, valid alignments, HISAT2 does not guarantee that
the N alignments reported are the best possible in terms of alignment score.
Still, this mode can be effective and fast in situations where the user cares
more about whether a read aligns (or aligns a certain number of times) than
where exactly it originated.</p>

<h2 id="alignment-summary">Alignment summary</h2>

<p>When HISAT2 finishes running, it prints messages summarizing what happened. 
These messages are printed to the “standard error” (“stderr”) filehandle.  For
datasets consisting of unpaired reads, the summary might look like this:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>20000 reads; of these:
  20000 (100.00%) were unpaired; of these:
    1247 (6.24%) aligned 0 times
    18739 (93.69%) aligned exactly 1 time
    14 (0.07%) aligned &gt;1 times
93.77% overall alignment rate
</code></pre></div></div>

<p>For datasets consisting of pairs, the summary might look like this:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>10000 reads; of these:
  10000 (100.00%) were paired; of these:
    650 (6.50%) aligned concordantly 0 times
    8823 (88.23%) aligned concordantly exactly 1 time
    527 (5.27%) aligned concordantly &gt;1 times
    ----
    650 pairs aligned concordantly 0 times; of these:
      34 (5.23%) aligned discordantly 1 time
    ----
    616 pairs aligned 0 times concordantly or discordantly; of these:
      1232 mates make up the pairs; of these:
        660 (53.57%) aligned 0 times
        571 (46.35%) aligned exactly 1 time
        1 (0.08%) aligned &gt;1 times
96.70% overall alignment rate
</code></pre></div></div>

<p>The indentation indicates how subtotals relate to totals.</p>

<h2 id="wrapper">Wrapper</h2>

<p>The <code class="highlighter-rouge">hisat2</code>, <code class="highlighter-rouge">hisat2-build</code> and <code class="highlighter-rouge">hisat2-inspect</code> executables are actually 
wrapper scripts that call binary programs as appropriate.  The wrappers shield
users from having to distinguish between “small” and “large” index formats,
discussed briefly in the following section.  Also, the <code class="highlighter-rouge">hisat2</code> wrapper
provides some key functionality, like the ability to handle compressed inputs,
and the functionality for <a href="#hisat2-options-un"><code class="highlighter-rouge">--un</code></a>, <a href="#hisat2-options-al"><code class="highlighter-rouge">--al</code></a> and related options.</p>

<p>It is recommended that you always run the hisat2 wrappers and not run the
binaries directly.</p>

<h2 id="small-and-large-indexes">Small and large indexes</h2>

<p><code class="highlighter-rouge">hisat2-build</code> can index reference genomes of any size.  For genomes less than
about 4 billion nucleotides in length, <code class="highlighter-rouge">hisat2-build</code> builds a “small” index
using 32-bit numbers in various parts of the index.  When the genome is longer,
<code class="highlighter-rouge">hisat2-build</code> builds a “large” index using 64-bit numbers.  Small indexes are
stored in files with the <code class="highlighter-rouge">.ht2</code> extension, and large indexes are stored in
files with the <code class="highlighter-rouge">.ht2l</code> extension.  The user need not worry about whether a
particular index is small or large; the wrapper scripts will automatically build
and use the appropriate index.</p>

<h2 id="performance-tuning">Performance tuning</h2>

<ol>
  <li>
    <p>If your computer has multiple processors/cores, use <code class="highlighter-rouge">-p</code></p>

    <p>The <a href="#hisat2-options-p"><code class="highlighter-rouge">-p</code></a> option causes HISAT2 to launch a specified number of parallel
search threads.  Each thread runs on a different processor/core and all
threads find alignments in parallel, increasing alignment throughput by
approximately a multiple of the number of threads (though in practice,
speedup is somewhat worse than linear).</p>
  </li>
</ol>

<h2 id="command-line">Command Line</h2>

<h3 id="setting-function-options">Setting function options</h3>

<p>Some HISAT2 options specify a function rather than an individual number or
setting.  In these cases the user specifies three parameters: (a) a function
type <code class="highlighter-rouge">F</code>, (b) a constant term <code class="highlighter-rouge">B</code>, and (c) a coefficient <code class="highlighter-rouge">A</code>.  The available
function types are constant (<code class="highlighter-rouge">C</code>), linear (<code class="highlighter-rouge">L</code>), square-root (<code class="highlighter-rouge">S</code>), and natural
log (<code class="highlighter-rouge">G</code>). The parameters are specified as <code class="highlighter-rouge">F,B,A</code> - that is, the function type,
the constant term, and the coefficient are separated by commas with no
whitespace.  The constant term and coefficient may be negative and/or
floating-point numbers.</p>

<p>For example, if the function specification is <code class="highlighter-rouge">L,-0.4,-0.6</code>, then the function
defined is:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>f(x) = -0.4 + -0.6 * x
</code></pre></div></div>

<p>If the function specification is <code class="highlighter-rouge">G,1,5.4</code>, then the function defined is:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>f(x) = 1.0 + 5.4 * ln(x)
</code></pre></div></div>

<p>See the documentation for the option in question to learn what the parameter <code class="highlighter-rouge">x</code>
is for.  For example, in the case if the <a href="#hisat2-options-score-min"><code class="highlighter-rouge">--score-min</code></a> option, the function
<code class="highlighter-rouge">f(x)</code> sets the minimum alignment score necessary for an alignment to be
considered valid, and <code class="highlighter-rouge">x</code> is the read length.</p>

<h3 id="usage">Usage</h3>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>hisat2 [options]* -x &lt;hisat2-idx&gt; {-1 &lt;m1&gt; -2 &lt;m2&gt; | -U &lt;r&gt; | --sra-acc &lt;SRA accession number&gt;} [-S &lt;hit&gt;]
</code></pre></div></div>

<h3 id="main-arguments">Main arguments</h3>

<ul id="hisat2-options-x">
  <li>-x <code class="highlighter-rouge">&lt;hisat2-idx&gt;</code><br />
The basename of the index for the reference genome.  The basename is the name of
any of the index files up to but not including the final <code class="highlighter-rouge">.1.ht2</code> / etc.
<code class="highlighter-rouge">hisat2</code> looks for the specified index first in the current directory,
then in the directory specified in the <code class="highlighter-rouge">HISAT2_INDEXES</code> environment variable.</li>
</ul>

<ul id="hisat2-options-1">
  <li>-1 <code class="highlighter-rouge">&lt;m1&gt;</code><br />
Comma-separated list of files containing mate 1s (filename usually includes
<code class="highlighter-rouge">_1</code>), e.g. <code class="highlighter-rouge">-1 flyA_1.fq,flyB_1.fq</code>.  Sequences specified with this option must
correspond file-for-file and read-for-read with those specified in <code class="highlighter-rouge">&lt;m2&gt;</code>. Reads
may be a mix of different lengths. If <code class="highlighter-rouge">-</code> is specified, <code class="highlighter-rouge">hisat2</code> will read the
mate 1s from the “standard in” or “stdin” filehandle.</li>
</ul>

<ul id="hisat2-options-2">
  <li>-2 <code class="highlighter-rouge">&lt;m2&gt;</code><br />
Comma-separated list of files containing mate 2s (filename usually includes
<code class="highlighter-rouge">_2</code>), e.g. <code class="highlighter-rouge">-2 flyA_2.fq,flyB_2.fq</code>.  Sequences specified with this option must
correspond file-for-file and read-for-read with those specified in <code class="highlighter-rouge">&lt;m1&gt;</code>. Reads
may be a mix of different lengths. If <code class="highlighter-rouge">-</code> is specified, <code class="highlighter-rouge">hisat2</code> will read the
mate 2s from the “standard in” or “stdin” filehandle.</li>
</ul>

<ul id="hisat2-options-U">
  <li>-U <code class="highlighter-rouge">&lt;r&gt;</code><br />
Comma-separated list of files containing unpaired reads to be aligned, e.g.
<code class="highlighter-rouge">lane1.fq,lane2.fq,lane3.fq,lane4.fq</code>.  Reads may be a mix of different lengths.
If <code class="highlighter-rouge">-</code> is specified, <code class="highlighter-rouge">hisat2</code> gets the reads from the “standard in” or “stdin”
filehandle.</li>
</ul>

<ul id="hisat2-options-sra-acc">
  <li>--sra-acc <code class="highlighter-rouge">&lt;SRA accession number&gt;</code><br />
Comma-separated list of SRA accession numbers, e.g. <code class="highlighter-rouge">--sra-acc SRR353653,SRR353654</code>.
Information about read types is available at <code>http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?sp=runinfo&amp;acc=<b>sra-acc</b>&amp;retmode=xml</code>,
where <b>sra-acc</b> is SRA accession number.  If users run HISAT2 on a computer cluster, it is recommended to disable SRA-related caching (see the instruction at <a href="https://github.com/ncbi/sra-tools/wiki/Toolkit-Configuration">SRA-MANUAL</a>).</li>
</ul>

<ul id="hisat2-options-S">
  <li>-S <code class="highlighter-rouge">&lt;hit&gt;</code><br />
File to write SAM alignments to.  By default, alignments are written to the
“standard out” or “stdout” filehandle (i.e. the console).</li>
</ul>

<h3 id="options">Options</h3>

<h4 id="input-options">Input options</h4>

<ul id="hisat2-options-q">
  <li>-q<br />
Reads (specified with <code class="highlighter-rouge">&lt;m1&gt;</code>, <code class="highlighter-rouge">&lt;m2&gt;</code>, <code class="highlighter-rouge">&lt;s&gt;</code>) are FASTQ files.  FASTQ files
usually have extension <code class="highlighter-rouge">.fq</code> or <code class="highlighter-rouge">.fastq</code>.  FASTQ is the default format.  See
also: <a href="#hisat2-options-solexa-quals"><code class="highlighter-rouge">--solexa-quals</code></a> and <a href="#hisat2-options-int-quals"><code class="highlighter-rouge">--int-quals</code></a>.</li>
</ul>

<ul id="hisat2-options-qseq">
  <li>--qseq<br />
Reads (specified with <code class="highlighter-rouge">&lt;m1&gt;</code>, <code class="highlighter-rouge">&lt;m2&gt;</code>, <code class="highlighter-rouge">&lt;s&gt;</code>) are QSEQ files.  QSEQ files usually
end in <code class="highlighter-rouge">_qseq.txt</code>.  See also: <a href="#hisat2-options-solexa-quals"><code class="highlighter-rouge">--solexa-quals</code></a> and <a href="#hisat2-options-int-quals"><code class="highlighter-rouge">--int-quals</code></a>.</li>
</ul>

<ul id="hisat2-options-f">
  <li>-f<br />
Reads (specified with <code class="highlighter-rouge">&lt;m1&gt;</code>, <code class="highlighter-rouge">&lt;m2&gt;</code>, <code class="highlighter-rouge">&lt;s&gt;</code>) are FASTA files.  FASTA files
usually have extension <code class="highlighter-rouge">.fa</code>, <code class="highlighter-rouge">.fasta</code>, <code class="highlighter-rouge">.mfa</code>, <code class="highlighter-rouge">.fna</code> or similar.  FASTA files
do not have a way of specifying quality values, so when <code class="highlighter-rouge">-f</code> is set, the result
is as if <code class="highlighter-rouge">--ignore-quals</code> is also set.</li>
</ul>

<ul id="hisat2-options-r">
  <li>-r<br />
Reads (specified with <code class="highlighter-rouge">&lt;m1&gt;</code>, <code class="highlighter-rouge">&lt;m2&gt;</code>, <code class="highlighter-rouge">&lt;s&gt;</code>) are files with one input sequence
per line, without any other information (no read names, no qualities).  When
<code class="highlighter-rouge">-r</code> is set, the result is as if <code class="highlighter-rouge">--ignore-quals</code> is also set.</li>
</ul>

<ul id="hisat2-options-c">
  <li>-c<br />
The read sequences are given on command line.  I.e. <code class="highlighter-rouge">&lt;m1&gt;</code>, <code class="highlighter-rouge">&lt;m2&gt;</code> and
<code class="highlighter-rouge">&lt;singles&gt;</code> are comma-separated lists of reads rather than lists of read files.
There is no way to specify read names or qualities, so <code class="highlighter-rouge">-c</code> also implies
<code class="highlighter-rouge">--ignore-quals</code>.</li>
</ul>

<ul id="hisat2-options-s">
  <li>-s/--skip <code class="highlighter-rouge">&lt;int&gt;</code><br />
Skip (i.e. do not align) the first <code class="highlighter-rouge">&lt;int&gt;</code> reads or pairs in the input.</li>
</ul>

<ul id="hisat2-options-u">
  <li>-u/--qupto <code class="highlighter-rouge">&lt;int&gt;</code><br />
Align the first <code class="highlighter-rouge">&lt;int&gt;</code> reads or read pairs from the input (after the
<a href="#hisat2-options-s"><code class="highlighter-rouge">-s</code>/<code class="highlighter-rouge">--skip</code></a> reads or pairs have been skipped), then stop.  Default: no limit.</li>
</ul>

<ul id="hisat2-options-5">
  <li>-5/--trim5 <code class="highlighter-rouge">&lt;int&gt;</code><br />
Trim <code class="highlighter-rouge">&lt;int&gt;</code> bases from 5’ (left) end of each read before alignment (default: 0).</li>
</ul>

<ul id="hisat2-options-3">
  <li>-3/--trim3 <code class="highlighter-rouge">&lt;int&gt;</code><br />
Trim <code class="highlighter-rouge">&lt;int&gt;</code> bases from 3’ (right) end of each read before alignment (default: 0).</li>
</ul>

<ul id="hisat2-options-phred33-quals">
  <li>--phred33<br />
Input qualities are ASCII chars equal to the <a href="http://en.wikipedia.org/wiki/Phred_quality_score">Phred quality</a> plus 33.  This is
also called the “Phred+33” encoding, which is used by the very latest Illumina
pipelines.</li>
</ul>

<ul id="hisat2-options-phred64-quals">
  <li>--phred64<br />
Input qualities are ASCII chars equal to the <a href="http://en.wikipedia.org/wiki/Phred_quality_score">Phred quality</a> plus 64.  This is
also called the “Phred+64” encoding.</li>
</ul>

<ul id="hisat2-options-solexa-quals">
  <li>--solexa-quals<br />
Convert input qualities from <a href="http://en.wikipedia.org/wiki/Phred_quality_score">Solexa</a> (which can be negative) to
<a href="http://en.wikipedia.org/wiki/Phred_quality_score">Phred</a> (which can’t).  This scheme was used in older Illumina GA
Pipeline versions (prior to 1.3).  Default: off.</li>
</ul>

<ul id="hisat2-options-int-quals">
  <li>--int-quals<br />
Quality values are represented in the read input file as space-separated ASCII
integers, e.g., <code class="highlighter-rouge">40 40 30 40</code>…, rather than ASCII characters, e.g., <code class="highlighter-rouge">II?I</code>….
Integers are treated as being on the <a href="http://en.wikipedia.org/wiki/Phred_quality_score">Phred quality</a> scale unless
<a href="#hisat2-options-solexa-quals"><code class="highlighter-rouge">--solexa-quals</code></a> is also specified. Default: off.</li>
</ul>

<h4 id="alignment-options">Alignment options</h4>

<ul id="hisat2-options-n-ceil">
  <li>--n-ceil <code class="highlighter-rouge">&lt;func&gt;</code><br />
Sets a function governing the maximum number of ambiguous characters (usually
<code class="highlighter-rouge">N</code>s and/or <code class="highlighter-rouge">.</code>s) allowed in a read as a function of read length.  For instance,
specifying <code class="highlighter-rouge">-L,0,0.15</code> sets the N-ceiling function <code class="highlighter-rouge">f</code> to <code class="highlighter-rouge">f(x) = 0 + 0.15 * x</code>,
where x is the read length.  See also: [setting function options].  Reads
exceeding this ceiling are <a href="#filtering">filtered out</a>.  Default: <code class="highlighter-rouge">L,0,0.15</code>.</li>
</ul>

<ul id="hisat2-options-ignore-quals">
  <li>--ignore-quals<br />
When calculating a mismatch penalty, always consider the quality value at the
mismatched position to be the highest possible, regardless of the actual value. 
I.e. input is treated as though all quality values are high.  This is also the
default behavior when the input doesn’t specify quality values (e.g. in <a href="#hisat2-options-f"><code class="highlighter-rouge">-f</code></a>,
<a href="#hisat2-options-r"><code class="highlighter-rouge">-r</code></a>, or <a href="#hisat2-options-c"><code class="highlighter-rouge">-c</code></a> modes).</li>
</ul>

<ul id="hisat2-options-nofw">
  <li>--nofw/--norc<br />
If <code class="highlighter-rouge">--nofw</code> is specified, <code class="highlighter-rouge">hisat2</code> will not attempt to align unpaired reads to
the forward (Watson) reference strand.  If <code class="highlighter-rouge">--norc</code> is specified, <code class="highlighter-rouge">hisat2</code> will
not attempt to align unpaired reads against the reverse-complement (Crick)
reference strand. In paired-end mode, <code class="highlighter-rouge">--nofw</code> and <code class="highlighter-rouge">--norc</code> pertain to the
fragments; i.e. specifying <code class="highlighter-rouge">--nofw</code> causes <code class="highlighter-rouge">hisat2</code> to explore only those
paired-end configurations corresponding to fragments from the reverse-complement
(Crick) strand.  Default: both strands enabled.</li>
</ul>

<h4 id="scoring-options">Scoring options</h4>

<ul id="hisat2-options-mp">
  <li>--mp <code class="highlighter-rouge">MX,MN</code><br />
Sets the maximum (<code class="highlighter-rouge">MX</code>) and minimum (<code class="highlighter-rouge">MN</code>) mismatch penalties, both integers.  A
number less than or equal to <code class="highlighter-rouge">MX</code> and greater than or equal to <code class="highlighter-rouge">MN</code> is
subtracted from the alignment score for each position where a read character
aligns to a reference character, the characters do not match, and neither is an
<code class="highlighter-rouge">N</code>.  If <a href="#hisat2-options-ignore-quals"><code class="highlighter-rouge">--ignore-quals</code></a> is specified, the number subtracted quals <code class="highlighter-rouge">MX</code>.
Otherwise, the number subtracted is <code class="highlighter-rouge">MN + floor( (MX-MN)(MIN(Q, 40.0)/40.0) )</code>
where Q is the Phred quality value.  Default: <code class="highlighter-rouge">MX</code> = 6, <code class="highlighter-rouge">MN</code> = 2.</li>
</ul>

<ul id="hisat2-options-sp">
  <li>--sp <code class="highlighter-rouge">MX,MN</code><br />
Sets the maximum (<code class="highlighter-rouge">MX</code>) and minimum (<code class="highlighter-rouge">MN</code>) penalties for soft-clipping per base, 
both integers. A number less than or equal to <code class="highlighter-rouge">MX</code> and greater than or equal to <code class="highlighter-rouge">MN</code> is
subtracted from the alignment score for each position.
The number subtracted is <code class="highlighter-rouge">MN + floor( (MX-MN)(MIN(Q, 40.0)/40.0) )</code>
where Q is the Phred quality value.  Default: <code class="highlighter-rouge">MX</code> = 2, <code class="highlighter-rouge">MN</code> = 1.</li>
</ul>

<ul id="hisat2-options-no-softclip">
  <li>--no-softclip<br />
Disallow soft-clipping.</li>
</ul>

<ul id="hisat2-options-np">
  <li>--np <code class="highlighter-rouge">&lt;int&gt;</code><br />
Sets penalty for positions where the read, reference, or both, contain an
ambiguous character such as <code class="highlighter-rouge">N</code>.  Default: 1.</li>
</ul>

<ul id="hisat2-options-rdg">
  <li>--rdg <code class="highlighter-rouge">&lt;int1&gt;,&lt;int2&gt;</code><br />
Sets the read gap open (<code class="highlighter-rouge">&lt;int1&gt;</code>) and extend (<code class="highlighter-rouge">&lt;int2&gt;</code>) penalties.  A read gap of
length N gets a penalty of <code class="highlighter-rouge">&lt;int1&gt;</code> + N * <code class="highlighter-rouge">&lt;int2&gt;</code>.  Default: 5, 3.</li>
</ul>

<ul id="hisat2-options-rfg">
  <li>--rfg <code class="highlighter-rouge">&lt;int1&gt;,&lt;int2&gt;</code><br />
Sets the reference gap open (<code class="highlighter-rouge">&lt;int1&gt;</code>) and extend (<code class="highlighter-rouge">&lt;int2&gt;</code>) penalties.  A
reference gap of length N gets a penalty of <code class="highlighter-rouge">&lt;int1&gt;</code> + N * <code class="highlighter-rouge">&lt;int2&gt;</code>.  Default:
5, 3.</li>
</ul>

<ul id="hisat2-options-score-min">
  <li>--score-min <code class="highlighter-rouge">&lt;func&gt;</code><br />
Sets a function governing the minimum alignment score needed for an alignment to
be considered “valid” (i.e. good enough to report).  This is a function of read
length. For instance, specifying <code class="highlighter-rouge">L,0,-0.6</code> sets the minimum-score function <code class="highlighter-rouge">f</code>
to <code class="highlighter-rouge">f(x) = 0 + -0.6 * x</code>, where <code class="highlighter-rouge">x</code> is the read length.  See also: [setting function options].  The default is <code class="highlighter-rouge">L,0,-0.2</code>.</li>
</ul>

<h4 id="spliced-alignment-options">Spliced alignment options</h4>

<ul id="hisat2-options-pen-cansplice">
  <li>--pen-cansplice <code class="highlighter-rouge">&lt;int&gt;</code><br />
Sets the penalty for each pair of canonical splice sites (e.g. GT/AG). Default: 0.</li>
</ul>

<ul id="hisat2-options-pen-noncansplice">
  <li>--pen-noncansplice <code class="highlighter-rouge">&lt;int&gt;</code><br />
Sets the penalty for each pair of non-canonical splice sites (e.g. non-GT/AG). Default: 12.</li>
</ul>

<ul id="hisat2-options-pen-canintronlen">
  <li>--pen-canintronlen <code class="highlighter-rouge">&lt;func&gt;</code><br />
Sets the penalty for long introns with canonical splice sites so that alignments with shorter introns are preferred
to those with longer ones.  Default: G,-8,1</li>
</ul>

<ul id="hisat2-options-pen-noncanintronlen">
  <li>--pen-noncanintronlen <code class="highlighter-rouge">&lt;func&gt;</code><br />
Sets the penalty for long introns with noncanonical splice sites so that alignments with shorter introns are preferred
to those with longer ones.  Default: G,-8,1</li>
</ul>

<ul id="hisat2-options-min-intronlen">
  <li>--min-intronlen <code class="highlighter-rouge">&lt;int&gt;</code><br />
Sets minimum intron length. Default: 20</li>
</ul>

<ul id="hisat2-options-max-intronlen">
  <li>--max-intronlen <code class="highlighter-rouge">&lt;int&gt;</code><br />
Sets maximum intron length. Default: 500000</li>
</ul>

<ul id="hisat2-options-known-splicesite-infile">
  <li>--known-splicesite-infile <code class="highlighter-rouge">&lt;path&gt;</code><br />
With this mode, you can provide a list of known splice sites, which HISAT2 makes use of to align reads with small anchors. <br />
You can create such a list using <code class="highlighter-rouge">python hisat2_extract_splice_sites.py genes.gtf &gt; splicesites.txt</code>,
where <code class="highlighter-rouge">hisat2_extract_splice_sites.py</code> is included in the HISAT2 package, <code class="highlighter-rouge">genes.gtf</code> is a gene annotation file,
and <code class="highlighter-rouge">splicesites.txt</code> is a list of splice sites with which you provide HISAT2 in this mode.
Note that it is better to use indexes built using annotated transcripts (such as <i>genome_tran</i> or <i>genome_snp_tran</i>), which works better
than using this option.  It has no effect to provide splice sites that are already included in the indexes.</li>
</ul>

<ul id="hisat2-options-novel-splicesite-outfile">
  <li>--novel-splicesite-outfile <code class="highlighter-rouge">&lt;path&gt;</code><br />
In this mode, HISAT2 reports a list of splice sites in the file <path>:
</path>    <blockquote>
      <p>chromosome name <code class="highlighter-rouge">&lt;tab&gt;</code> genomic position of the flanking base on the left side of an intron <code class="highlighter-rouge">&lt;tab&gt;</code> genomic position of the flanking base on the right <code class="highlighter-rouge">&lt;tab&gt;</code> strand (+, -, and .)</p>
    </blockquote>

    <p>’.’ indicates an unknown strand for non-canonical splice sites.</p>
  </li>
</ul>

<ul id="hisat2-options-novel-splicesite-infile">
  <li>--novel-splicesite-infile <code class="highlighter-rouge">&lt;path&gt;</code><br />
With this mode, you can provide a list of novel splice sites that were generated from the above option “–novel-splicesite-outfile”.</li>
</ul>

<ul id="hisat2-options-no-temp-splicesite">
  <li>--no-temp-splicesite<br />
HISAT2, by default, makes use of splice sites found by earlier reads to align later reads in the same run,
in particular, reads with small anchors (&lt;= 15 bp).<br />
The option disables this default alignment strategy.</li>
</ul>

<ul id="hisat2-options-no-spliced-alignment">
  <li>--no-spliced-alignment<br />
Disable spliced alignment.</li>
</ul>

<ul id="hisat2-options-rna-strandness">
  <li>--rna-strandness <code class="highlighter-rouge">&lt;string&gt;</code><br />
Specify strand-specific information: the default is unstranded.<br />
For single-end reads, use F or R.
    <blockquote>
      <p>‘F’ means a read corresponds to a transcript.<br />
‘R’ means a read corresponds to the reverse complemented counterpart of a transcript.</p>
    </blockquote>

    <p>For paired-end reads, use either FR or RF.<br />
With this option being used, every read alignment will have an XS attribute tag:</p>
    <blockquote>
      <p>’+’ means a read belongs to a transcript on ‘+’ strand of genome.<br />
‘-‘ means a read belongs to a transcript on ‘-‘ strand of genome.</p>
    </blockquote>

    <p>(TopHat has a similar option, –library-type option, where fr-firststrand corresponds to R and RF; fr-secondstrand corresponds to F and FR.)</p>
  </li>
</ul>

<ul id="hisat2-options-tmo">
  <li>--tmo/–transcriptome-mapping-only<br />
Report only those alignments within known transcripts.</li>
</ul>

<ul id="hisat2-options-dta">
  <li>--dta/--downstream-transcriptome-assembly<br />
Report alignments tailored for transcript assemblers including StringTie.
With this option, HISAT2 requires longer anchor lengths for de novo discovery of splice sites.
This leads to fewer alignments with short-anchors, 
which helps transcript assemblers improve significantly in computation and memory usage.</li>
</ul>

<ul id="hisat2-options-dta-cufflinks">
  <li>--dta-cufflinks<br />
Report alignments tailored specifically for Cufflinks. In addition to what HISAT2 does with the above option (–dta),
With this option, HISAT2 looks for novel splice sites with three signals (GT/AG, GC/AG, AT/AC), but all user-provided splice sites are used irrespective of their signals.
HISAT2 produces an optional field, XS:A:[+-], for every spliced alignment.</li>
</ul>

<ul id="hisat2-options-avoid-pseudogene">
  <li>--avoid-pseudogene<br />
Try to avoid aligning reads to pseudogenes.  Note this option is experimental and needs further investigation.</li>
</ul>

<ul id="hisat2-options-no-templatelen-adjustment">
  <li>--no-templatelen-adjustment<br />
Disables template length adjustment for RNA-seq reads.</li>
</ul>

<h4 id="reporting-options">Reporting options</h4>

<ul id="hisat2-options-k">
  <li>-k <code class="highlighter-rouge">&lt;int&gt;</code><br />
It searches for at most <code class="highlighter-rouge">&lt;int&gt;</code> distinct, primary alignments for each read.
Primary alignments mean alignments whose alignment score is equal or higher than any other alignments.
The search terminates when it can’t find more distinct valid alignments, or when it
finds <code class="highlighter-rouge">&lt;int&gt;</code>, whichever happens first. The alignment score for a paired-end
alignment equals the sum of the alignment scores of the individual mates. Each
reported read or pair alignment beyond the first has the SAM ‘secondary’ bit
(which equals 256) set in its FLAGS field.  For reads that have more than
<code class="highlighter-rouge">&lt;int&gt;</code> distinct, valid alignments, <code class="highlighter-rouge">hisat2</code> does not guarantee that the
<code class="highlighter-rouge">&lt;int&gt;</code> alignments reported are the best possible in terms of alignment score. Default: 5 (HFM) or 10 (HGFM)
    <p />

    <p>Note: HISAT2 is not designed with large values for <code class="highlighter-rouge">-k</code> in mind, and when
aligning reads to long, repetitive genomes large <code class="highlighter-rouge">-k</code> can be very, very slow.</p>
  </li>
</ul>

<ul id="hisat2-options-max-seeds">
  <li>--max-seeds <code class="highlighter-rouge">&lt;int&gt;</code><br />
HISAT2, like other aligners, uses seed-and-extend approaches.  HISAT2 tries to extend seeds to full-length alignments. In HISAT2, –max-seeds is used to control the maximum number of seeds that will be extended. HISAT2 extends up to these many seeds and skips the rest of the seeds. Large values for <code class="highlighter-rouge">--max-seeds</code> may improve alignment sensitivity, but HISAT2 is not designed with large values for <code class="highlighter-rouge">--max-seeds</code> in mind, and when aligning reads to long, repetitive genomes large <code class="highlighter-rouge">--max-seeds</code> can be very, very slow. The default value is the maximum of 5 and the value that comes with<code class="highlighter-rouge">-k</code>.</li>
</ul>

<ul id="hisat2-options-secondary">
  <li>--secondary<br />
Report secondary alignments.</li>
</ul>

<h4 id="paired-end-options">Paired-end options</h4>

<ul id="hisat2-options-I">
  <li>-I/--minins <code class="highlighter-rouge">&lt;int&gt;</code><br />
The minimum fragment length for valid paired-end alignments.This option is valid only with <a href="#hisat2-options-no-spliced-alignment"><code class="highlighter-rouge">--no-spliced-alignment</code></a>.
E.g. if <code class="highlighter-rouge">-I 60</code> is specified and a paired-end alignment consists of two 20-bp alignments in the
appropriate orientation with a 20-bp gap between them, that alignment is
considered valid (as long as <a href="#hisat2-options-X"><code class="highlighter-rouge">-X</code></a> is also satisfied). A 19-bp gap would not
be valid in that case.  If trimming options <a href="#hisat2-options-3"><code class="highlighter-rouge">-3</code></a> or <a href="#hisat2-options-5"><code class="highlighter-rouge">-5</code></a> are also used, the
<a href="#hisat2-options-I"><code class="highlighter-rouge">-I</code></a> constraint is applied with respect to the untrimmed mates.
    <p />

    <p>The larger the difference between <a href="#hisat2-options-I"><code class="highlighter-rouge">-I</code></a> and <a href="#hisat2-options-X"><code class="highlighter-rouge">-X</code></a>, the slower HISAT2 will
run.  This is because larger differences between <a href="#hisat2-options-I"><code class="highlighter-rouge">-I</code></a> and <a href="#hisat2-options-X"><code class="highlighter-rouge">-X</code></a> require that
HISAT2 scan a larger window to determine if a concordant alignment exists.
For typical fragment length ranges (200 to 400 nucleotides), HISAT2 is very
efficient.</p>
    <p />

    <p>Default: 0 (essentially imposing no minimum)</p>
  </li>
</ul>

<ul id="hisat2-options-X">
  <li>-X/--maxins <code class="highlighter-rouge">&lt;int&gt;</code><br />
The maximum fragment length for valid paired-end alignments. This option is valid only with --no-spliced-alignment.
E.g. if <code class="highlighter-rouge">-X 100</code> is specified and a paired-end alignment consists of two 20-bp alignments in the
proper orientation with a 60-bp gap between them, that alignment is considered
valid (as long as <a href="#hisat2-options-I"><code class="highlighter-rouge">-I</code></a> is also satisfied).  A 61-bp gap would not be valid in
that case.  If trimming options <a href="#hisat2-options-3"><code class="highlighter-rouge">-3</code></a> or <a href="#hisat2-options-5"><code class="highlighter-rouge">-5</code></a> are also used, the <code class="highlighter-rouge">-X</code>
constraint is applied with respect to the untrimmed mates, not the trimmed
mates.
    <p />

    <p>The larger the difference between <a href="#hisat2-options-I"><code class="highlighter-rouge">-I</code></a> and <a href="#hisat2-options-X"><code class="highlighter-rouge">-X</code></a>, the slower HISAT2 will
run.  This is because larger differences between <a href="#hisat2-options-I"><code class="highlighter-rouge">-I</code></a> and <a href="#hisat2-options-X"><code class="highlighter-rouge">-X</code></a> require that
HISAT2 scan a larger window to determine if a concordant alignment exists.
For typical fragment length ranges (200 to 400 nucleotides), HISAT2 is very
efficient.</p>
    <p />

    <p>Default: 500.</p>
  </li>
</ul>

<ul id="hisat2-options-fr">
  <li>--fr/--rf/--ff<br />
The upstream/downstream mate orientations for a valid paired-end alignment
against the forward reference strand.  E.g., if <code class="highlighter-rouge">--fr</code> is specified and there is
a candidate paired-end alignment where mate 1 appears upstream of the reverse
complement of mate 2 and the fragment length constraints (<a href="#hisat2-options-I"><code class="highlighter-rouge">-I</code></a> and <a href="#hisat2-options-X"><code class="highlighter-rouge">-X</code></a>) are
met, that alignment is valid.  Also, if mate 2 appears upstream of the reverse
complement of mate 1 and all other constraints are met, that too is valid.
<code class="highlighter-rouge">--rf</code> likewise requires that an upstream mate1 be reverse-complemented and a
downstream mate2 be forward-oriented. <code class="highlighter-rouge">--ff</code> requires both an upstream mate 1
and a downstream mate 2 to be forward-oriented.  Default: <code class="highlighter-rouge">--fr</code> (appropriate
for Illumina’s Paired-end Sequencing Assay).</li>
</ul>

<ul id="hisat2-options-no-mixed">
  <li>--no-mixed<br />
By default, when <code class="highlighter-rouge">hisat2</code> cannot find a concordant or discordant alignment for
a pair, it then tries to find alignments for the individual mates.  This option
disables that behavior.</li>
</ul>

<ul id="hisat2-options-no-discordant">
  <li>--no-discordant<br />
By default, <code class="highlighter-rouge">hisat2</code> looks for discordant alignments if it cannot find any
concordant alignments.  A discordant alignment is an alignment where both mates
align uniquely, but that does not satisfy the paired-end constraints
(<a href="#hisat2-options-fr"><code class="highlighter-rouge">--fr</code>/<code class="highlighter-rouge">--rf</code>/<code class="highlighter-rouge">--ff</code></a>, <a href="#hisat2-options-I"><code class="highlighter-rouge">-I</code></a>, <a href="#hisat2-options-X"><code class="highlighter-rouge">-X</code></a>).  This option disables that behavior.</li>
</ul>

<h4 id="output-options">Output options</h4>

<ul id="hisat2-options-t">
  <li>-t/--time<br />
Print the wall-clock time required to load the index files and align the reads. 
This is printed to the “standard error” (“stderr”) filehandle.  Default: off.</li>
</ul>

<ul id="hisat2-options-un">
  <li>--un <code class="highlighter-rouge">&lt;path&gt;</code>, --un-gz <code class="highlighter-rouge">&lt;path&gt;</code>, --un-bz2 <code class="highlighter-rouge">&lt;path&gt;</code><br />
Write unpaired reads that fail to align to file at <code class="highlighter-rouge">&lt;path&gt;</code>.  These reads
correspond to the SAM records with the FLAGS <code class="highlighter-rouge">0x4</code> bit set and neither the
<code class="highlighter-rouge">0x40</code> nor <code class="highlighter-rouge">0x80</code> bits set.  If <code class="highlighter-rouge">--un-gz</code> is specified, output will be gzip
compressed. If <code class="highlighter-rouge">--un-bz2</code> is specified, output will be bzip2 compressed.  Reads
written in this way will appear exactly as they did in the input file, without
any modification (same sequence, same name, same quality string, same quality
encoding).  Reads will not necessarily appear in the same order as they did in
the input.</li>
</ul>

<ul id="hisat2-options-al">
  <li>--al <code class="highlighter-rouge">&lt;path&gt;</code>, --al-gz <code class="highlighter-rouge">&lt;path&gt;</code>, --al-bz2 <code class="highlighter-rouge">&lt;path&gt;</code><br />
Write unpaired reads that align at least once to file at <code class="highlighter-rouge">&lt;path&gt;</code>.  These reads
correspond to the SAM records with the FLAGS <code class="highlighter-rouge">0x4</code>, <code class="highlighter-rouge">0x40</code>, and <code class="highlighter-rouge">0x80</code> bits
unset.  If <code class="highlighter-rouge">--al-gz</code> is specified, output will be gzip compressed. If <code class="highlighter-rouge">--al-bz2</code>
is specified, output will be bzip2 compressed.  Reads written in this way will
appear exactly as they did in the input file, without any modification (same
sequence, same name, same quality string, same quality encoding).  Reads will
not necessarily appear in the same order as they did in the input.</li>
</ul>

<ul id="hisat2-options-un-conc">
  <li>--un-conc <code class="highlighter-rouge">&lt;path&gt;</code>, --un-conc-gz <code class="highlighter-rouge">&lt;path&gt;</code>, --un-conc-bz2 <code class="highlighter-rouge">&lt;path&gt;</code><br />
Write paired-end reads that fail to align concordantly to file(s) at <code class="highlighter-rouge">&lt;path&gt;</code>.
These reads correspond to the SAM records with the FLAGS <code class="highlighter-rouge">0x4</code> bit set and
either the <code class="highlighter-rouge">0x40</code> or <code class="highlighter-rouge">0x80</code> bit set (depending on whether it’s mate #1 or #2).
<code class="highlighter-rouge">.1</code> and <code class="highlighter-rouge">.2</code> strings are added to the filename to distinguish which file
contains mate #1 and mate #2.  If a percent symbol, <code class="highlighter-rouge">%</code>, is used in <code class="highlighter-rouge">&lt;path&gt;</code>,
the percent symbol is replaced with <code class="highlighter-rouge">1</code> or <code class="highlighter-rouge">2</code> to make the per-mate filenames.
Otherwise, <code class="highlighter-rouge">.1</code> or <code class="highlighter-rouge">.2</code> are added before the final dot in <code class="highlighter-rouge">&lt;path&gt;</code> to make the
per-mate filenames.  Reads written in this way will appear exactly as they did
in the input files, without any modification (same sequence, same name, same
quality string, same quality encoding).  Reads will not necessarily appear in
the same order as they did in the inputs.</li>
</ul>

<ul id="hisat2-options-al-conc">
  <li>--al-conc <code class="highlighter-rouge">&lt;path&gt;</code>, --al-conc-gz <code class="highlighter-rouge">&lt;path&gt;</code>, --al-conc-bz2 <code class="highlighter-rouge">&lt;path&gt;</code><br />
Write paired-end reads that align concordantly at least once to file(s) at
<code class="highlighter-rouge">&lt;path&gt;</code>. These reads correspond to the SAM records with the FLAGS <code class="highlighter-rouge">0x4</code> bit
unset and either the <code class="highlighter-rouge">0x40</code> or <code class="highlighter-rouge">0x80</code> bit set (depending on whether it’s mate #1
or #2). <code class="highlighter-rouge">.1</code> and <code class="highlighter-rouge">.2</code> strings are added to the filename to distinguish which
file contains mate #1 and mate #2.  If a percent symbol, <code class="highlighter-rouge">%</code>, is used in
<code class="highlighter-rouge">&lt;path&gt;</code>, the percent symbol is replaced with <code class="highlighter-rouge">1</code> or <code class="highlighter-rouge">2</code> to make the per-mate
filenames. Otherwise, <code class="highlighter-rouge">.1</code> or <code class="highlighter-rouge">.2</code> are added before the final dot in <code class="highlighter-rouge">&lt;path&gt;</code> to
make the per-mate filenames.  Reads written in this way will appear exactly as
they did in the input files, without any modification (same sequence, same name,
same quality string, same quality encoding).  Reads will not necessarily appear
in the same order as they did in the inputs.</li>
</ul>

<ul id="hisat2-options-quiet">
  <li>--quiet<br />
Print nothing besides alignments and serious errors.</li>
</ul>

<ul id="hisat2-options-summary-file">
  <li>--summary-file<br />
Print alignment summary to this file.</li>
</ul>

<ul id="hisat2-options-new-summary">
  <li>--new-summary<br />
Print alignment summary in a new style, which is more machine-friendly.</li>
</ul>

<ul id="hisat2-options-met-file">
  <li>--met-file <code class="highlighter-rouge">&lt;path&gt;</code><br />
Write <code class="highlighter-rouge">hisat2</code> metrics to file <code class="highlighter-rouge">&lt;path&gt;</code>.  Having alignment metric can be useful
for debugging certain problems, especially performance issues.  See also:
<a href="#hisat2-options-met"><code class="highlighter-rouge">--met</code></a>.  Default: metrics disabled.</li>
</ul>

<ul id="hisat2-options-met-stderr">
  <li>--met-stderr<br />
Write <code class="highlighter-rouge">hisat2</code> metrics to the “standard error” (“stderr”) filehandle.  This is
not mutually exclusive with <a href="#hisat2-options-met-file"><code class="highlighter-rouge">--met-file</code></a>.  Having alignment metric can be
useful for debugging certain problems, especially performance issues.  See also:
<a href="#hisat2-options-met"><code class="highlighter-rouge">--met</code></a>.  Default: metrics disabled.</li>
</ul>

<ul id="hisat2-options-met">
  <li>--met <code class="highlighter-rouge">&lt;int&gt;</code><br />
Write a new <code class="highlighter-rouge">hisat2</code> metrics record every <code class="highlighter-rouge">&lt;int&gt;</code> seconds.  Only matters if
either <a href="#hisat2-options-met-stderr"><code class="highlighter-rouge">--met-stderr</code></a> or <a href="#hisat2-options-met-file"><code class="highlighter-rouge">--met-file</code></a> are specified.  Default: 1.</li>
</ul>

<h4 id="sam-options">SAM options</h4>

<ul id="hisat2-options-no-unal">
  <li>--no-unal<br />
Suppress SAM records for reads that failed to align.</li>
</ul>

<ul id="hisat2-options-no-hd">
  <li>--no-hd<br />
Suppress SAM header lines (starting with <code class="highlighter-rouge">@</code>).</li>
</ul>

<ul id="hisat2-options-no-sq">
  <li>--no-sq<br />
Suppress <code class="highlighter-rouge">@SQ</code> SAM header lines.</li>
</ul>

<ul id="hisat2-options-rg-id">
  <li>--rg-id <code class="highlighter-rouge">&lt;text&gt;</code><br />
Set the read group ID to <code class="highlighter-rouge">&lt;text&gt;</code>.  This causes the SAM <code class="highlighter-rouge">@RG</code> header line to be
printed, with <code class="highlighter-rouge">&lt;text&gt;</code> as the value associated with the <code class="highlighter-rouge">ID:</code> tag.  It also
causes the <code class="highlighter-rouge">RG:Z:</code> extra field to be attached to each SAM output record, with
value set to <code class="highlighter-rouge">&lt;text&gt;</code>.</li>
</ul>

<ul id="hisat2-options-rg">
  <li>--rg <code class="highlighter-rouge">&lt;text&gt;</code><br />
Add <code class="highlighter-rouge">&lt;text&gt;</code> (usually of the form <code class="highlighter-rouge">TAG:VAL</code>, e.g. <code class="highlighter-rouge">SM:Pool1</code>) as a field on the
<code class="highlighter-rouge">@RG</code> header line.  Note: in order for the <code class="highlighter-rouge">@RG</code> line to appear, <a href="#hisat2-options-rg-id"><code class="highlighter-rouge">--rg-id</code></a>
must also be specified.  This is because the <code class="highlighter-rouge">ID</code> tag is required by the <a href="http://samtools.sourceforge.net/SAM1.pdf">SAM
Spec</a>.  Specify <code class="highlighter-rouge">--rg</code> multiple times to set multiple fields.  See the
<a href="http://samtools.sourceforge.net/SAM1.pdf">SAM Spec</a> for details about what fields are legal.</li>
</ul>

<ul id="hisat2-remove-chrname">
  <li>--remove-chrname<br />
Remove ‘chr’ from reference names in alignment (e.g., chr18 to 18)</li>
</ul>

<ul id="hisat2-options-add-chrname">
  <li>--add-chrname<br />
Add ‘chr’ to reference names in alignment (e.g., 18 to chr18)</li>
</ul>

<ul id="hisat2-options-omit-sec-seq">
  <li>--omit-sec-seq<br />
When printing secondary alignments, HISAT2 by default will write out the <code class="highlighter-rouge">SEQ</code>
and <code class="highlighter-rouge">QUAL</code> strings.  Specifying this option causes HISAT2 to print an asterisk
in those fields instead.</li>
</ul>

<h4 id="performance-options">Performance options</h4>

<ul id="hisat2-options-o">
  <li>-o/--offrate <code class="highlighter-rouge">&lt;int&gt;</code><br />
Override the offrate of the index with <code class="highlighter-rouge">&lt;int&gt;</code>.  If <code class="highlighter-rouge">&lt;int&gt;</code> is greater
than the offrate used to build the index, then some row markings are
discarded when the index is read into memory.  This reduces the memory
footprint of the aligner but requires more time to calculate text
offsets.  <code class="highlighter-rouge">&lt;int&gt;</code> must be greater than the value used to build the
index.</li>
</ul>

<ul id="hisat2-options-p">
  <li>-p/--threads <code class="highlighter-rouge">NTHREADS</code><br />
Launch <code class="highlighter-rouge">NTHREADS</code> parallel search threads (default: 1).  Threads will run on
separate processors/cores and synchronize when parsing reads and outputting
alignments.  Searching for alignments is highly parallel, and speedup is close
to linear.  Increasing <code class="highlighter-rouge">-p</code> increases HISAT2’s memory footprint. E.g. when
aligning to a human genome index, increasing <code class="highlighter-rouge">-p</code> from 1 to 8 increases the
memory footprint by a few hundred megabytes.  This option is only available if
<code class="highlighter-rouge">bowtie</code> is linked with the <code class="highlighter-rouge">pthreads</code> library.</li>
</ul>

<ul id="hisat2-options-reorder">
  <li>--reorder<br />
Guarantees that output SAM records are printed in an order corresponding to the
order of the reads in the original input file, even when <a href="#hisat2-options-p"><code class="highlighter-rouge">-p</code></a> is set greater
than 1.  Specifying <code class="highlighter-rouge">--reorder</code> and setting <a href="#hisat2-options-p"><code class="highlighter-rouge">-p</code></a> greater than 1 causes HISAT2
to run somewhat slower and use somewhat more memory then if <code class="highlighter-rouge">--reorder</code> were
not specified.  Has no effect if <a href="#hisat2-options-p"><code class="highlighter-rouge">-p</code></a> is set to 1, since output order will
naturally correspond to input order in that case.</li>
</ul>

<ul id="hisat2-options-mm">
  <li>--mm<br />
Use memory-mapped I/O to load the index, rather than typical file I/O.
Memory-mapping allows many concurrent <code class="highlighter-rouge">bowtie</code> processes on the same computer to
share the same memory image of the index (i.e. you pay the memory overhead just
once).  This facilitates memory-efficient parallelization of <code class="highlighter-rouge">bowtie</code> in
situations where using <a href="#hisat2-options-p"><code class="highlighter-rouge">-p</code></a> is not possible or not preferable.</li>
</ul>

<h4 id="other-options">Other options</h4>

<ul id="hisat2-options-qc-filter">
  <li>--qc-filter<br />
Filter out reads for which the QSEQ filter field is non-zero.  Only has an
effect when read format is <a href="#hisat2-options-qseq"><code class="highlighter-rouge">--qseq</code></a>.  Default: off.</li>
</ul>

<ul id="hisat2-options-seed">
  <li>--seed <code class="highlighter-rouge">&lt;int&gt;</code><br />
Use <code class="highlighter-rouge">&lt;int&gt;</code> as the seed for pseudo-random number generator.  Default: 0.</li>
</ul>

<ul id="hisat2-options-non-deterministic">
  <li>--non-deterministic<br />
Normally, HISAT2 re-initializes its pseudo-random generator for each read.  It
seeds the generator with a number derived from (a) the read name, (b) the
nucleotide sequence, (c) the quality sequence, (d) the value of the <a href="#hisat2-options-seed"><code class="highlighter-rouge">--seed</code></a>
option.  This means that if two reads are identical (same name, same
nucleotides, same qualities) HISAT2 will find and report the same alignment(s)
for both, even if there was ambiguity.  When <code class="highlighter-rouge">--non-deterministic</code> is specified,
HISAT2 re-initializes its pseudo-random generator for each read using the
current time.  This means that HISAT2 will not necessarily report the same
alignment for two identical reads.  This is counter-intuitive for some users,
but might be more appropriate in situations where the input consists of many
identical reads.</li>
</ul>

<ul id="hisat2-options-version">
  <li>--version<br />
Print version information and quit.</li>
</ul>

<ul id="hisat2-options-h">
  <li>-h/--help<br />
Print usage information and quit.</li>
</ul>

<h2 id="sam-output">SAM output</h2>

<p>Following is a brief description of the <a href="http://samtools.sourceforge.net/SAM1.pdf">SAM</a> format as output by <code class="highlighter-rouge">hisat2</code>. 
For more details, see the <a href="http://samtools.sourceforge.net/SAM1.pdf">SAM format specification</a>.</p>

<p>By default, <code class="highlighter-rouge">hisat2</code> prints a SAM header with <code class="highlighter-rouge">@HD</code>, <code class="highlighter-rouge">@SQ</code> and <code class="highlighter-rouge">@PG</code> lines. 
When one or more <a href="#hisat2-options-rg"><code class="highlighter-rouge">--rg</code></a> arguments are specified, <code class="highlighter-rouge">hisat2</code> will also print
an <code class="highlighter-rouge">@RG</code> line that includes all user-specified <a href="#hisat2-options-rg"><code class="highlighter-rouge">--rg</code></a> tokens separated by
tabs.</p>

<p>Each subsequent line describes an alignment or, if the read failed to align, a
read.  Each line is a collection of at least 12 fields separated by tabs; from
left to right, the fields are:</p>

<ol>
  <li>Name of read that aligned. 
Note that the <a href="http://samtools.sourceforge.net/SAM1.pdf">SAM specification</a> disallows whitespace in the read name.
If the read name contains any whitespace characters, HISAT2 will truncate
the name at the first whitespace character.  This is similar to the
behavior of other tools.</li>
  <li>Sum of all applicable flags.<br />
Flags relevant to HISAT2 are
    <ul>
      <li>1: The read is one of a pair</li>
      <li>2: The alignment is one end of a proper paired-end alignment</li>
      <li>4: The read has no reported alignments</li>
      <li>8: The read is one of a pair and has no reported alignments</li>
      <li>16: The alignment is to the reverse reference strand</li>
      <li>32: The other mate in the paired-end alignment is aligned to the reverse reference strand</li>
      <li>64: The read is mate 1 in a pair</li>
      <li>128: The read is mate 2 in a pair</li>
    </ul>
    <p>Thus, an unpaired read that aligns to the reverse reference strand
will have flag 16.  A paired-end read that aligns and is the first
mate in the pair will have flag 83 (= 64 + 16 + 2 + 1).</p>
  </li>
  <li>Name of reference sequence where alignment occurs</li>
  <li>1-based offset into the forward reference strand where leftmost
character of the alignment occurs</li>
  <li>Mapping quality.  Mapping quality of HISAT2</li>
  <li>CIGAR string representation of alignment</li>
  <li>Name of reference sequence where mate’s alignment occurs.  Set to <code class="highlighter-rouge">=</code> if the
mate’s reference sequence is the same as this alignment’s, or <code class="highlighter-rouge">*</code> if there is no
mate.</li>
  <li>1-based offset into the forward reference strand where leftmost character of
the mate’s alignment occurs.  Offset is 0 if there is no mate.</li>
  <li>Inferred fragment length.  Size is negative if the mate’s alignment occurs
upstream of this alignment.  Size is 0 if the mates did not align concordantly.
However, size is non-0 if the mates aligned discordantly to the same
chromosome.</li>
  <li>Read sequence (reverse-complemented if aligned to the reverse strand)</li>
  <li>ASCII-encoded read qualities (reverse-complemented if the read aligned to
the reverse strand).  The encoded quality values are on the <a href="http://en.wikipedia.org/wiki/Phred_quality_score">Phred quality</a>
scale and the encoding is ASCII-offset by 33 (ASCII char <code class="highlighter-rouge">!</code>), similarly to a
<a href="http://en.wikipedia.org/wiki/FASTQ_format">FASTQ</a> file.</li>
  <li>Optional fields.  Fields are tab-separated.  <code class="highlighter-rouge">hisat2</code> outputs zero or more
of these optional fields for each alignment, depending on the type of the
alignment:
    <ul>
      <li id="hisat2-opt-fields-as"><code class="highlighter-rouge">AS:i:&lt;N&gt;</code> : Alignment score.  Can be negative.  Only present if SAM record is for
an aligned read.</li>
      <li id="hisat2-opt-fields-xs"><code class="highlighter-rouge">ZS:i:&lt;N&gt;</code> : Alignment score for the best-scoring alignment found other than the
alignment reported.  Can be negative.  Only present if the SAM record is
for an aligned read and more than one alignment was found for the read.
Note that, when the read is part of a concordantly-aligned pair, this score
could be greater than <a href="#hisat2-opt-fields-as"><code class="highlighter-rouge">AS:i</code></a>.</li>
      <li id="hisat2-opt-fields-ys"><code class="highlighter-rouge">YS:i:&lt;N&gt;</code> : Alignment score for opposite mate in the paired-end alignment.  Only present
if the SAM record is for a read that aligned as part of a paired-end
alignment.</li>
      <li id="hisat2-opt-fields-xn"><code class="highlighter-rouge">XN:i:&lt;N&gt;</code> : The number of ambiguous bases in the reference covering this alignment. 
Only present if SAM record is for an aligned read.</li>
      <li id="hisat2-opt-fields-xm"><code class="highlighter-rouge">XM:i:&lt;N&gt;</code> : The number of mismatches in the alignment.  Only present if SAM record is
for an aligned read.</li>
      <li id="hisat2-opt-fields-xo"><code class="highlighter-rouge">XO:i:&lt;N&gt;</code> : The number of gap opens, for both read and reference gaps, in the alignment.
Only present if SAM record is for an aligned read.</li>
      <li id="hisat2-opt-fields-xg"><code class="highlighter-rouge">XG:i:&lt;N&gt;</code> : The number of gap extensions, for both read and reference gaps, in the
alignment. Only present if SAM record is for an aligned read.</li>
      <li id="hisat2-opt-fields-nm"><code class="highlighter-rouge">NM:i:&lt;N&gt;</code> : The edit distance; that is, the minimal number of one-nucleotide edits
(substitutions, insertions and deletions) needed to transform the read
string into the reference string.  Only present if SAM record is for an
aligned read.</li>
      <li id="hisat2-opt-fields-yf"><code class="highlighter-rouge">YF:Z:&lt;S&gt;</code> : String indicating reason why the read was filtered out.  See also:
[Filtering].  Only appears for reads that were filtered out.</li>
      <li id="hisat2-opt-fields-yt"><code class="highlighter-rouge">YT:Z:&lt;S&gt;</code> : Value of <code class="highlighter-rouge">UU</code> indicates the read was not part of a pair.  Value of <code class="highlighter-rouge">CP</code>
indicates the read was part of a pair and the pair aligned concordantly.
Value of <code class="highlighter-rouge">DP</code> indicates the read was part of a pair and the pair aligned
discordantly.  Value of <code class="highlighter-rouge">UP</code> indicates the read was part of a pair but the
pair failed to aligned either concordantly or discordantly.</li>
      <li id="hisat2-opt-fields-md"><code class="highlighter-rouge">MD:Z:&lt;S&gt;</code> : A string representation of the mismatched reference bases in the alignment. 
See <a href="http://samtools.sourceforge.net/SAM1.pdf">SAM</a> format specification for details.  Only present if SAM record is
for an aligned read.</li>
      <li id="hisat2-opt-fields-xs"><code class="highlighter-rouge">XS:A:&lt;A&gt;</code> : Values of <code class="highlighter-rouge">+</code> and <code class="highlighter-rouge">-</code> indicate the read is mapped to transcripts on sense and anti-sense
strands, respectively.  Spliced alignments need to have this field, which is required in Cufflinks and StringTie.<br />
We can report this field for the canonical-splice site (GT/AG), but not for non-canonical splice sites.
You can direct HISAT2 not to output such alignments (involving non-canonical splice sites)  using “–pen-noncansplice 1000000”.</li>
      <li id="hisat2-opt-fields-nh"><code class="highlighter-rouge">NH:i:&lt;N&gt;</code> : The number of mapped locations for the read or the pair.</li>
      <li id="hisat2-opt-fields-Zs"><code class="highlighter-rouge">Zs:Z:&lt;S&gt;</code> : When the alignment of a read involves SNPs that are in the index, this option is used to indicate where exactly the read involves the SNPs.
This optional field is similar to the above MD:Z field.
For example, <code class="highlighter-rouge">Zs:Z:1|S|rs3747203,97|S|rs16990981</code> indicates the second base of the read corresponds to a known SNP (ID: rs3747203).
97 bases after the third base (the base after the second one), the read at 100th base involves another known SNP (ID: rs16990981).
‘S’ indicates a single nucleotide polymorphism.  ‘D’ and ‘I’ indicate a deletion and an insertion, respectively.</li>
    </ul>
  </li>
</ol>

<h1 id="the-hisat2-build-indexer">The <code class="highlighter-rouge">hisat2-build</code> indexer</h1>

<p><code class="highlighter-rouge">hisat2-build</code> builds a HISAT2 index from a set of DNA sequences.
<code class="highlighter-rouge">hisat2-build</code> outputs a set of 6 files with suffixes <code class="highlighter-rouge">.1.ht2</code>, <code class="highlighter-rouge">.2.ht2</code>,
<code class="highlighter-rouge">.3.ht2</code>, <code class="highlighter-rouge">.4.ht2</code>, <code class="highlighter-rouge">.5.ht2</code>, <code class="highlighter-rouge">.6.ht2</code>, <code class="highlighter-rouge">.7.ht2</code>, and <code class="highlighter-rouge">.8.ht2</code>.  In the case of a large 
index these suffixes will have a <code class="highlighter-rouge">ht2l</code> termination.  These files together
constitute the index: they are all that is needed to align reads to that
reference.  The original sequence FASTA files are no longer used by HISAT2
once the index is built.</p>

<p>Use of Karkkainen’s <a href="http://portal.acm.org/citation.cfm?id=1314852">blockwise algorithm</a> allows <code class="highlighter-rouge">hisat2-build</code> to trade off
between running time and memory usage. <code class="highlighter-rouge">hisat2-build</code> has three options
governing how it makes this trade: [<code class="highlighter-rouge">-p</code>/<code class="highlighter-rouge">--packed</code>], <a href="#hisat2-build-options-bmax"><code class="highlighter-rouge">--bmax</code></a>/<a href="#hisat2-build-options-bmaxdivn"><code class="highlighter-rouge">--bmaxdivn</code></a>,
and <a href="#hisat2-build-options-dcv"><code class="highlighter-rouge">--dcv</code></a>.  By default, <code class="highlighter-rouge">hisat2-build</code> will automatically search for the
settings that yield the best running time without exhausting memory.  This
behavior can be disabled using the <a href="#hisat2-build-options-a"><code class="highlighter-rouge">-a</code>/<code class="highlighter-rouge">--noauto</code></a> option.</p>

<p>The indexer provides options pertaining to the “shape” of the index, e.g.
<a href="#hisat2-build-options-o"><code class="highlighter-rouge">--offrate</code></a> governs the fraction of <a href="http://en.wikipedia.org/wiki/Burrows-Wheeler_transform">Burrows-Wheeler</a>
rows that are “marked” (i.e., the density of the suffix-array sample; see the
original <a href="http://en.wikipedia.org/wiki/FM-index">FM Index</a> paper for details).  All of these options are potentially
profitable trade-offs depending on the application.  They have been set to
defaults that are reasonable for most cases according to our experiments.  See
<a href="#performance-tuning">Performance tuning</a> for details.</p>

<p><code class="highlighter-rouge">hisat2-build</code> can generate either <a href="#small-and-large-indexes">small or large indexes</a>.  The wrapper
will decide which based on the length of the input genome.  If the reference
does not exceed 4 billion characters but a large index is preferred,  the user
can specify <a href="#hisat2-build-options-large-index"><code class="highlighter-rouge">--large-index</code></a> to force <code class="highlighter-rouge">hisat2-build</code> to build a large index
instead.</p>

<p>The HISAT2 index is based on the <a href="http://en.wikipedia.org/wiki/FM-index">FM Index</a> of Ferragina and Manzini, which in
turn is based on the <a href="http://en.wikipedia.org/wiki/Burrows-Wheeler_transform">Burrows-Wheeler</a> transform.  The algorithm used to build
the index is based on the <a href="http://portal.acm.org/citation.cfm?id=1314852">blockwise algorithm</a> of Karkkainen.</p>

<h2 id="command-line-1">Command Line</h2>

<p>Usage:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>hisat2-build [options]* &lt;reference_in&gt; &lt;ht2_base&gt;
</code></pre></div></div>

<h3 id="notes">Notes</h3>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>If you use --snp, --ss, and/or --exon, hisat2-build will need about 200GB RAM for the human genome size as index building involves a graph construction. 
Otherwise, you will be able to build an index on your desktop with 8GB RAM.
</code></pre></div></div>

<h3 id="main-arguments-1">Main arguments</h3>

<ul id="hisat2-build-options-ref">
  <li><code class="highlighter-rouge">&lt;reference_in&gt;</code><br />
A comma-separated list of FASTA files containing the reference sequences to be
aligned to, or, if <a href="#hisat2-build-options-c"><code class="highlighter-rouge">-c</code></a> is specified, the sequences
themselves. E.g., <code class="highlighter-rouge">&lt;reference_in&gt;</code> might be <code class="highlighter-rouge">chr1.fa,chr2.fa,chrX.fa,chrY.fa</code>,
or, if <a href="#hisat2-build-options-c"><code class="highlighter-rouge">-c</code></a> is specified, this might be
<code class="highlighter-rouge">GGTCATCCT,ACGGGTCGT,CCGTTCTATGCGGCTTA</code>.</li>
</ul>

<ul id="hisat2-build-options-base">
  <li><code class="highlighter-rouge">&lt;ht2_base&gt;</code><br />
The basename of the index files to write.  By default, <code class="highlighter-rouge">hisat2-build</code> writes
files named <code class="highlighter-rouge">NAME.1.ht2</code>, <code class="highlighter-rouge">NAME.2.ht2</code>, <code class="highlighter-rouge">NAME.3.ht2</code>, <code class="highlighter-rouge">NAME.4.ht2</code>,
<code class="highlighter-rouge">NAME.5.ht2</code>, <code class="highlighter-rouge">NAME.6.ht2</code>, <code class="highlighter-rouge">NAME.7.ht2</code>, and <code class="highlighter-rouge">NAME.8.ht2</code> where <code class="highlighter-rouge">NAME</code> is <code class="highlighter-rouge">&lt;ht2_base&gt;</code>.</li>
</ul>

<h3 id="options-1">Options</h3>

<ul id="hisat2-build-options-f">
  <li>-f<br />
The reference input files (specified as <code class="highlighter-rouge">&lt;reference_in&gt;</code>) are FASTA files
(usually having extension <code class="highlighter-rouge">.fa</code>, <code class="highlighter-rouge">.mfa</code>, <code class="highlighter-rouge">.fna</code> or similar).</li>
</ul>

<ul id="hisat2-build-options-c">
  <li>-c<br />
The reference sequences are given on the command line.  I.e. <code class="highlighter-rouge">&lt;reference_in&gt;</code> is
a comma-separated list of sequences rather than a list of FASTA files.</li>
</ul>

<ul id="hisat2-build-options-large-index">
  <li>--large-index<br />
Force <code class="highlighter-rouge">hisat2-build</code> to build a <a href="#small-and-large-indexes">large index</a>, even if the reference is less
than ~ 4 billion nucleotides long.</li>
</ul>

<ul id="hisat2-build-options-a">
  <li>-a/--noauto<br />
Disable the default behavior whereby <code class="highlighter-rouge">hisat2-build</code> automatically selects
values for the <a href="#hisat2-build-options-bmax"><code class="highlighter-rouge">--bmax</code></a>, <a href="#hisat2-build-options-dcv"><code class="highlighter-rouge">--dcv</code></a> and [<code class="highlighter-rouge">--packed</code>] parameters according to
available memory.  Instead, user may specify values for those parameters.  If
memory is exhausted during indexing, an error message will be printed; it is up
to the user to try new parameters.</li>
</ul>

<ul id="hisat2-build-options-bmax">
  <li>--bmax <code class="highlighter-rouge">&lt;int&gt;</code><br />
The maximum number of suffixes allowed in a block.  Allowing more suffixes per
block makes indexing faster, but increases peak memory usage.  Setting this
option overrides any previous setting for <a href="#hisat2-build-options-bmax"><code class="highlighter-rouge">--bmax</code></a>, or <a href="#hisat2-build-options-bmaxdivn"><code class="highlighter-rouge">--bmaxdivn</code></a>. 
Default (in terms of the <a href="#hisat2-build-options-bmaxdivn"><code class="highlighter-rouge">--bmaxdivn</code></a> parameter) is <a href="#hisat2-build-options-bmaxdivn"><code class="highlighter-rouge">--bmaxdivn</code></a> 4.  This is
configured automatically by default; use <a href="#hisat2-build-options-a"><code class="highlighter-rouge">-a</code>/<code class="highlighter-rouge">--noauto</code></a> to configure manually.</li>
</ul>

<ul id="hisat2-build-options-bmaxdivn">
  <li>--bmaxdivn <code class="highlighter-rouge">&lt;int&gt;</code><br />
The maximum number of suffixes allowed in a block, expressed as a fraction of
the length of the reference.  Setting this option overrides any previous setting
for <a href="#hisat2-build-options-bmax"><code class="highlighter-rouge">--bmax</code></a>, or <a href="#hisat2-build-options-bmaxdivn"><code class="highlighter-rouge">--bmaxdivn</code></a>.  Default: <a href="#hisat2-build-options-bmaxdivn"><code class="highlighter-rouge">--bmaxdivn</code></a> 4.  This is
configured automatically by default; use <a href="#hisat2-build-options-a"><code class="highlighter-rouge">-a</code>/<code class="highlighter-rouge">--noauto</code></a> to configure manually.</li>
</ul>

<ul id="hisat2-build-options-dcv">
  <li>--dcv <code class="highlighter-rouge">&lt;int&gt;</code><br />
Use <code class="highlighter-rouge">&lt;int&gt;</code> as the period for the difference-cover sample.  A larger period
yields less memory overhead, but may make suffix sorting slower, especially if
repeats are present.  Must be a power of 2 no greater than 4096.  Default: 1024.
This is configured automatically by default; use <a href="#hisat2-build-options-a"><code class="highlighter-rouge">-a</code>/<code class="highlighter-rouge">--noauto</code></a> to configure
manually.</li>
</ul>

<ul id="hisat2-build-options-nodc">
  <li>--nodc<br />
Disable use of the difference-cover sample.  Suffix sorting becomes
quadratic-time in the worst case (where the worst case is an extremely
repetitive reference).  Default: off.</li>
</ul>

<ul id="hisat2-build-options-r">
  <li>-r/--noref<br />
Do not build the <code class="highlighter-rouge">NAME.3.ht2</code> and <code class="highlighter-rouge">NAME.4.ht2</code> portions of the index, which
contain a bitpacked version of the reference sequences and are used for
paired-end alignment.</li>
</ul>

<ul id="hisat2-build-options-3">
  <li>-3/--justref<br />
Build only the <code class="highlighter-rouge">NAME.3.ht2</code> and <code class="highlighter-rouge">NAME.4.ht2</code> portions of the index, which
contain a bitpacked version of the reference sequences and are used for
paired-end alignment.</li>
</ul>

<ul id="hisat2-build-options-o">
  <li>-o/--offrate <code class="highlighter-rouge">&lt;int&gt;</code><br />
To map alignments back to positions on the reference sequences, it’s necessary
to annotate (“mark”) some or all of the <a href="http://en.wikipedia.org/wiki/Burrows-Wheeler_transform">Burrows-Wheeler</a> rows with their
corresponding location on the genome. 
<a href="#hisat2-build-options-o"><code class="highlighter-rouge">-o</code>/<code class="highlighter-rouge">--offrate</code></a> governs how many rows get marked:
the indexer will mark every 2^<code class="highlighter-rouge">&lt;int&gt;</code> rows.  Marking more rows makes
reference-position lookups faster, but requires more memory to hold the
annotations at runtime.  The default is 4 (every 16th row is marked; for human
genome, annotations occupy about 680 megabytes).</li>
</ul>

<ul id="hisat2-build-options-t">
  <li>-t/--ftabchars <code class="highlighter-rouge">&lt;int&gt;</code><br />
The ftab is the lookup table used to calculate an initial <a href="http://en.wikipedia.org/wiki/Burrows-Wheeler_transform">Burrows-Wheeler</a>
range with respect to the first <code class="highlighter-rouge">&lt;int&gt;</code> characters of the query.  A larger
<code class="highlighter-rouge">&lt;int&gt;</code> yields a larger lookup table but faster query times.  The ftab has size
4^(<code class="highlighter-rouge">&lt;int&gt;</code>+1) bytes.  The default setting is 10 (ftab is 4MB).</li>
</ul>

<ul id="hisat2-build-options-localoffrate">
  <li>--localoffrate <code class="highlighter-rouge">&lt;int&gt;</code><br />
This option governs how many rows get marked in a local index:
the indexer will mark every 2^<code class="highlighter-rouge">&lt;int&gt;</code> rows.  Marking more rows makes
reference-position lookups faster, but requires more memory to hold the
annotations at runtime.  The default is 3 (every 8th row is marked,
this occupies about 16KB per local index).</li>
</ul>

<ul id="hisat2-build-options-localftabchars">
  <li>--localftabchars <code class="highlighter-rouge">&lt;int&gt;</code><br />
The local ftab is the lookup table in a local index.
The default setting is 6 (ftab is 8KB per local index).</li>
</ul>

<ul id="hisat2-build-options-p">
  <li>-p <code class="highlighter-rouge">&lt;int&gt;</code><br />
Launch <code class="highlighter-rouge">NTHREADS</code> parallel build threads (default: 1).</li>
</ul>

<ul id="hisat2-build-options-snp">
  <li>--snp <code class="highlighter-rouge">&lt;path&gt;</code><br />
Provide a list of SNPs (in the HISAT2’s own format) as follows (five columns).
    <blockquote>
      <p>SNP ID <code class="highlighter-rouge">&lt;tab&gt;</code> snp type (single, deletion, or insertion) <code class="highlighter-rouge">&lt;tab&gt;</code> chromosome name <code class="highlighter-rouge">&lt;tab&gt;</code> zero-offset based genomic position of a SNP <code class="highlighter-rouge">&lt;tab&gt;</code> alternative base (single), the length of SNP (deletion), or insertion sequence (insertion)</p>
    </blockquote>

    <p>For example,<br />
<code class="highlighter-rouge">rs58784443      single  13      18447947        T</code></p>

    <p>Use <code class="highlighter-rouge">hisat2_extract_snps_haplotypes_UCSC.py</code> (in the HISAT2 package) to extract SNPs and haplotypes from a dbSNP file (e.g. http://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/snp144Common.txt.gz).
or <code class="highlighter-rouge">hisat2_extract_snps_haplotypes_VCF.py</code> to extract SNPs and haplotypes from a VCF file (e.g. ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr22.phase3_shapeit2_mvncall_integrated_v3plus_nounphased.rsID.genotypes.GRCh38_dbSNP_no_SVs.vcf.gz).</p>
  </li>
</ul>

<ul id="hisat2-build-options-haplotype">
  <li>--haplotype <code class="highlighter-rouge">&lt;path&gt;</code><br />
Provide a list of haplotypes (in the HISAT2’s own format) as follows (five columns).
    <blockquote>
      <p>Haplotype ID <code class="highlighter-rouge">&lt;tab&gt;</code> chromosome name <code class="highlighter-rouge">&lt;tab&gt;</code> zero-offset based left coordinate of haplotype <code class="highlighter-rouge">&lt;tab&gt;</code> zero-offset based right coordinate of haplotype <code class="highlighter-rouge">&lt;tab&gt;</code> a comma separated list of SNP ids in the haplotype</p>
    </blockquote>

    <p>For example,<br />
<code class="highlighter-rouge">ht35    13      18446877        18446945        rs12381094,rs12381056,rs192016659,rs538569910</code></p>

    <p>See the above option, –snp, about how to extract haplotypes.  This option is not required, but haplotype information can keep the index construction from exploding and reduce the index size substantially.</p>
  </li>
</ul>

<ul id="hisat2-build-options-ss">
  <li>--ss <code class="highlighter-rouge">&lt;path&gt;</code><br />
Note this option should be used with the following –exon option.
Provide a list of splice sites (in the HISAT2’s own format) as follows (four columns).
    <blockquote>
      <p>chromosome name <code class="highlighter-rouge">&lt;tab&gt;</code> zero-offset based genomic position of the flanking base on the left side of an intron <code class="highlighter-rouge">&lt;tab&gt;</code> zero-offset based genomic position of the flanking base on the right <code class="highlighter-rouge">&lt;tab&gt;</code> strand</p>
    </blockquote>

    <p>Use <code class="highlighter-rouge">hisat2_extract_splice_sites.py</code> (in the HISAT2 package) to extract splice sites from a GTF file.</p>
  </li>
</ul>

<ul id="hisat2-build-options-exon">
  <li>--exon <code class="highlighter-rouge">&lt;path&gt;</code><br />
Note this option should be used with the above –ss option.
Provide a list of exons (in the HISAT2’s own format) as follows (three columns).
    <blockquote>
      <p>chromosome name <code class="highlighter-rouge">&lt;tab&gt;</code> zero-offset based left genomic position of an exon <code class="highlighter-rouge">&lt;tab&gt;</code> zero-offset based right genomic position of an exon</p>
    </blockquote>

    <p>Use <code class="highlighter-rouge">hisat2_extract_exons.py</code> (in the HISAT2 package) to extract exons from a GTF file.</p>
  </li>
</ul>

<ul id="hisat2-build-options-seed">
  <li>--seed <code class="highlighter-rouge">&lt;int&gt;</code><br />
Use <code class="highlighter-rouge">&lt;int&gt;</code> as the seed for pseudo-random number generator.</li>
</ul>

<ul id="hisat2-build-options-cutoff">
  <li>--cutoff <code class="highlighter-rouge">&lt;int&gt;</code><br />
Index only the first <code class="highlighter-rouge">&lt;int&gt;</code> bases of the reference sequences (cumulative across
sequences) and ignore the rest.</li>
</ul>

<ul id="hisat2-build-options-q">
  <li>-q/--quiet<br />
<code class="highlighter-rouge">hisat2-build</code> is verbose by default.  With this option <code class="highlighter-rouge">hisat2-build</code> will
print only error messages.</li>
</ul>

<ul id="hisat2-build-options-h">
  <li>-h/--help<br />
Print usage information and quit.</li>
</ul>

<ul id="hisat2-build-options-version">
  <li>--version<br />
Print version information and quit.</li>
</ul>

<h1 id="the-hisat2-inspect-index-inspector">The <code class="highlighter-rouge">hisat2-inspect</code> index inspector</h1>

<p><code class="highlighter-rouge">hisat2-inspect</code> extracts information from a HISAT2 index about what kind of
index it is and what reference sequences were used to build it. When run without
any options, the tool will output a FASTA file containing the sequences of the
original references (with all non-<code class="highlighter-rouge">A</code>/<code class="highlighter-rouge">C</code>/<code class="highlighter-rouge">G</code>/<code class="highlighter-rouge">T</code> characters converted to <code class="highlighter-rouge">N</code>s).
 It can also be used to extract just the reference sequence names using the
<a href="#hisat2-inspect-options-n"><code class="highlighter-rouge">-n</code>/<code class="highlighter-rouge">--names</code></a> option or a more verbose summary using the <a href="#hisat2-inspect-options-s"><code class="highlighter-rouge">-s</code>/<code class="highlighter-rouge">--summary</code></a>
option.</p>

<h2 id="command-line-2">Command Line</h2>

<p>Usage:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>hisat2-inspect [options]* &lt;ht2_base&gt;
</code></pre></div></div>

<h3 id="main-arguments-2">Main arguments</h3>

<ul id="hisat2-inspect-options-base">
  <li><code class="highlighter-rouge">&lt;ht2_base&gt;</code><br />
The basename of the index to be inspected.  The basename is name of any of the
index files but with the <code class="highlighter-rouge">.X.ht2</code> suffix omitted.
<code class="highlighter-rouge">hisat2-inspect</code> first looks in the current directory for the index files, then
in the directory specified in the <code class="highlighter-rouge">HISAT2_INDEXES</code> environment variable.</li>
</ul>

<h3 id="options-2">Options</h3>

<ul id="hisat2-inspect-options-a">
  <li>-a/--across <code class="highlighter-rouge">&lt;int&gt;</code><br />
When printing FASTA output, output a newline character every <code class="highlighter-rouge">&lt;int&gt;</code> bases
(default: 60).</li>
</ul>

<ul id="hisat2-inspect-options-n">
  <li>-n/--names<br />
Print reference sequence names, one per line, and quit.</li>
</ul>

<ul id="hisat2-inspect-options-s">
  <li>-s/--summary<br />
Print a summary that includes information about index settings, as well as the
names and lengths of the input sequences.  The summary has this format:
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  Colorspace	&lt;0 or 1&gt;
  SA-Sample	1 in &lt;sample&gt;
  FTab-Chars	&lt;chars&gt;
  Sequence-1	&lt;name&gt;	&lt;len&gt;
  Sequence-2	&lt;name&gt;	&lt;len&gt;
  ...
  Sequence-N	&lt;name&gt;	&lt;len&gt;
</code></pre></div>    </div>
    <p>Fields are separated by tabs.  Colorspace is always set to 0 for HISAT2.</p>
  </li>
</ul>

<ul id="hisat2-inspect-options-snp">
  <li>--snp<br />
Print SNPs, and quit.</li>
</ul>

<ul id="hisat2-inspect-options-ss">
  <li>--ss<br />
Print splice sites, and quit.</li>
</ul>

<ul id="hisat2-inspect-options-ss-all">
  <li>--ss-all<br />
Print splice sites including those not in the global index, and quit.</li>
</ul>

<ul id="hisat2-inspect-options-exon">
  <li>--exon<br />
Print exons, and quit.</li>
</ul>

<ul>
  <li>-v/--verbose<br />
Print verbose output (for debugging).</li>
</ul>
<ul>
  <li>--version<br />
Print version information and quit.</li>
</ul>
<ul>
  <li>-h/--help<br />
Print usage information and quit.</li>
</ul>

<h1 id="getting-started-with-hisat2">Getting started with HISAT2</h1>

<p>HISAT2 comes with some example files to get you started.  The example files
are not scientifically significant; these files will simply let you start running HISAT2 and
downstream tools right away.</p>

<p>First follow the manual instructions to <a href="#obtaining-hisat2">obtain HISAT2</a>.  Set the <code class="highlighter-rouge">HISAT2_HOME</code>
environment variable to point to the new HISAT2 directory containing the
<code class="highlighter-rouge">hisat2</code>, <code class="highlighter-rouge">hisat2-build</code> and <code class="highlighter-rouge">hisat2-inspect</code> binaries.  This is important,
as the <code class="highlighter-rouge">HISAT2_HOME</code> variable is used in the commands below to refer to that
directory.</p>

<h2 id="indexing-a-reference-genome">Indexing a reference genome</h2>

<p>To create an index for the genomic region (1 million bps from the human chromosome 22 between 20,000,000 and 20,999,999)
included with HISAT2, create a new temporary directory (it doesn’t matter where), change into that directory, and run:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$HISAT2_HOME/hisat2-build $HISAT2_HOME/example/reference/22_20-21M.fa --snp $HISAT2_HOME/example/reference/22_20-21M.snp 22_20-21M_snp
</code></pre></div></div>

<p>The command should print many lines of output then quit. When the command
completes, the current directory will contain ten new files that all start with
<code class="highlighter-rouge">22_20-21M_snp</code> and end with <code class="highlighter-rouge">.1.ht2</code>, <code class="highlighter-rouge">.2.ht2</code>, <code class="highlighter-rouge">.3.ht2</code>, <code class="highlighter-rouge">.4.ht2</code>, <code class="highlighter-rouge">.5.ht2</code>, <code class="highlighter-rouge">.6.ht2</code>,
<code class="highlighter-rouge">.7.ht2</code>, and <code class="highlighter-rouge">.8.ht2</code>.  These files constitute the index - you’re done!</p>

<p>You can use <code class="highlighter-rouge">hisat2-build</code> to create an index for a set of FASTA files obtained
from any source, including sites such as <a href="http://genome.ucsc.edu/cgi-bin/hgGateway">UCSC</a>, <a href="http://www.ncbi.nlm.nih.gov/sites/genome">NCBI</a>, and <a href="http://www.ensembl.org/">Ensembl</a>. When
indexing multiple FASTA files, specify all the files using commas to separate
file names.  For more details on how to create an index with <code class="highlighter-rouge">hisat2-build</code>,
see the <a href="#the-hisat2-build-indexer">manual section on index building</a>.  You may also want to bypass this
process by obtaining a pre-built index.</p>

<h2 id="aligning-example-reads">Aligning example reads</h2>

<p>Stay in the directory created in the previous step, which now contains the
<code class="highlighter-rouge">22_20-21M</code> index files.  Next, run:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$HISAT2_HOME/hisat2 -f -x $HISAT2_HOME/example/index/22_20-21M_snp -U $HISAT2_HOME/example/reads/reads_1.fa -S eg1.sam
</code></pre></div></div>

<p>This runs the HISAT2 aligner, which aligns a set of unpaired reads to the
genome region using the index generated in the previous step.
The alignment results in SAM format are written to the file <code class="highlighter-rouge">eg1.sam</code>, and a
short alignment summary is written to the console.  (Actually, the summary is
written to the “standard error” or “stderr” filehandle, which is typically
printed to the console.)</p>

<p>To see the first few lines of the SAM output, run:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>head eg1.sam
</code></pre></div></div>

<p>You will see something like this:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>@HD     VN:1.0  SO:unsorted
@SQ     SN:22:20000001-21000000 LN:1000000
@PG     ID:hisat2       PN:hisat2       VN:2.0.0-beta
1       0       22:20000001-21000000    397984  255     100M    *       0       0       GCCTGTGAGGGAGCCCCGGACCCGGTCAGAGCAGGAGCCTGGCCTGGGGCCAAGTTCACCTTATGGACTCTCTTCCCTGCCCTTCCAGGAGCAGCTCACT    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:100        YT:Z:UU NH:i:1
2       16      22:20000001-21000000    398131  255     100M    *       0       0       ATGACACACTGTACACACCAGGGGCCCTGTGCTCCCCAGGAAGAGGGCCCTCACTTGAAGCGGGGCCCGATGGCCGCCACGTGCCGGTTCATGCTCCCCT    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:80A19      YT:Z:UU NH:i:1  Zs:Z:80|S|rs576159895
3       16      22:20000001-21000000    398222  255     100M    *       0       0       TGCTCCCCTTGGCCCCGCCGATGTTCAGGGACATGGAGCGCTGCAGCAGGCTGGAGAAGATCTCCACTTGGTCAGAGCTGCAGTACTTGGCGATCTCAAA    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:16A83      YT:Z:UU NH:i:1  Zs:Z:16|S|rs2629364
4       16      22:20000001-21000000    398247  255     90M200N10M      *       0       0       CAGGGACATGGAGCGCTGCAGCAGGCTGGAGAAGATCTCCACTTGGTCAGAGCTGCAGTACTTGGCGATCTCAAACCGCTGCACCAGGAAGTCGATCCAG    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:100        YT:Z:UU XS:A:-  NH:i:1
5       16      22:20000001-21000000    398194  255     100M    *       0       0       GGCCCGATGGCCGCCACGTGCCGGTTCATGCTCCCCTTGGCCCCGCCGATGTTCAGGGACATGGAGCGCTGCAGCAGGCTGGAGAAGATCTCCACTTGGT    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:17A26A55   YT:Z:UU NH:i:1  Zs:Z:17|S|rs576159895,26|S|rs2629364
6       0       22:20000001-21000000    398069  255     100M    *       0       0       CAGGAGCAGCTCACTGAAATGTGTTCCCCGTCTACAGAAGTACCGTGATACACAGACGCCCCATGACACACTGTACACACCAGGGGCCCTGTGCTCCCCA    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:100        YT:Z:UU NH:i:1
7       0       22:20000001-21000000    397896  255     100M    *       0       0       GTGGAGTAGATCTTCTCGCGAAGCACATTGCAGATGGTTGCATTTGGAACCACATCGGCATGCAGGAGGGACAGCCCCAGGGTCAGCAGCCTGTGAGGGA    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:31G68      YT:Z:UU NH:i:1  Zs:Z:31|S|rs562662261
8       0       22:20000001-21000000    398150  255     100M    *       0       0       AGGGGCCCTGTGCTCCCCAGGAAGAGGGCCCTCACTTGAAGCGGGGCCCGATGGCCGCCACGTGCCGGTTCATGCTCCCCTTGGCCCCGCCGATGTTCAG    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:61A26A11   YT:Z:UU NH:i:1  Zs:Z:61|S|rs576159895,26|S|rs2629364
9       16      22:20000001-21000000    398329  255     8M200N92M       *       0       0       ACCAGGAAGTCGATCCAGATGTAGTGGGGGGTCACTTCGGGGGGACAGGGTTTGGGTTGACTTGCTTCCGAGGCAGCCAGGGGGTCTGCTTCCTTTATCT    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:100        YT:Z:UU XS:A:-  NH:i:1
10      16      22:20000001-21000000    398184  255     100M    *       0       0       CTTGAAGCGGGGCCCGATGGCCGCCACGTGCCGGTTCATGCTCCCCTTGGCCCCGCCGATGTTCAGGGACATGGAGCGCTGCAGCAGGCTGGAGAAGATC    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:27A26A45   YT:Z:UU NH:i:1  Zs:Z:27|S|rs576159895,26|S|rs2629364
</code></pre></div></div>

<p>The first few lines (beginning with <code class="highlighter-rouge">@</code>) are SAM header lines, and the rest of
the lines are SAM alignments, one line per read or mate.  See the <a href="#sam-output">HISAT2
manual section on SAM output</a> and the <a href="http://samtools.sourceforge.net/SAM1.pdf">SAM specification</a> for details about how
to interpret the SAM file format.</p>

<h2 id="paired-end-example">Paired-end example</h2>

<p>To align paired-end reads included with HISAT2, stay in the same directory and
run:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$HISAT2_HOME/hisat2 -f -x $HISAT2_HOME/example/index/22_20-21M_snp -1 $HISAT2_HOME/example/reads/reads_1.fa -2 $HISAT2_HOME/example/reads/reads_2.fa -S eg2.sam
</code></pre></div></div>

<p>This aligns a set of paired-end reads to the reference genome, with results
written to the file <code class="highlighter-rouge">eg2.sam</code>.</p>

<h2 id="using-samtoolsbcftools-downstream">Using SAMtools/BCFtools downstream</h2>

<p><a href="http://samtools.sourceforge.net">SAMtools</a> is a collection of tools for manipulating and analyzing SAM and BAM
alignment files.  <a href="http://samtools.sourceforge.net/mpileup.shtml">BCFtools</a> is a collection of tools for calling variants and
manipulating VCF and BCF files, and it is typically distributed with <a href="http://samtools.sourceforge.net">SAMtools</a>.
Using these tools together allows you to get from alignments in SAM format to
variant calls in VCF format.  This example assumes that <code class="highlighter-rouge">samtools</code> and
<code class="highlighter-rouge">bcftools</code> are installed and that the directories containing these binaries are
in your <a href="http://en.wikipedia.org/wiki/PATH_(variable)">PATH environment variable</a>.</p>

<p>Run the paired-end example:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$HISAT2_HOME/hisat -f -x $HISAT2_HOME/example/index/22_20-21M_snp -1 $HISAT2_HOME/example/reads/reads_1.fa -2 $HISAT2_HOME/example/reads/reads_2.fa -S eg2.sam
</code></pre></div></div>

<p>Use <code class="highlighter-rouge">samtools view</code> to convert the SAM file into a BAM file.  BAM is a the
binary format corresponding to the SAM text format.  Run:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools view -bS eg2.sam &gt; eg2.bam
</code></pre></div></div>

<p>Use <code class="highlighter-rouge">samtools sort</code> to convert the BAM file to a sorted BAM file. The following command requires samtools version 1.2 or higher.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools sort eg2.bam -o eg2.sorted.bam
</code></pre></div></div>

<p>We now have a sorted BAM file called <code class="highlighter-rouge">eg2.sorted.bam</code>. Sorted BAM is a useful
format because the alignments are (a) compressed, which is convenient for
long-term storage, and (b) sorted, which is convenient for variant discovery.
To generate variant calls in VCF format, run:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools mpileup -uf $HISAT2_HOME/example/reference/22_20-21M.fa eg2.sorted.bam | bcftools view -bvcg - &gt; eg2.raw.bcf
</code></pre></div></div>

<p>Then to view the variants, run:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bcftools view eg2.raw.bcf
</code></pre></div></div>

<p>See the official SAMtools guide to <a href="http://samtools.sourceforge.net/mpileup.shtml">Calling SNPs/INDELs with SAMtools/BCFtools</a>
for more details and variations on this process.</p>


        </section>
        <footer>
    
</footer>

    </article>
</div>

    </div>

    <aside class="site-aside">
        <div class="inner">
            <div class="block">
                <form action="/hisatgenotype/search">
                    <input type="search" id="search" name="q" placeholder="Search" />
                </form>
            </div>
            <ul>
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                <li><a class="page-link" href="/hisatgenotype/main/">Main</a></li>
                
                
                
                <li><a class="page-link" href="/hisatgenotype/about/">About</a></li>
                
                
                
                <li><a class="page-link" href="/hisatgenotype/tags/">Tags</a></li>
                
                
                
                <li><a class="page-link" href="/hisatgenotype/archives/">Archives</a></li>
                
                
                
                <li><a class="page-link" href="/hisatgenotype/manual/">Manual</a></li>
                
                
                
                <li><a class="page-link" href="/hisatgenotype/howto/">HOWTO</a></li>
                
                
            </ul>
            <ul class="icons">
                


<li>
    <a href="/hisatgenotype/feed.xml">
        <i class="fa fa-fw fa-rss"></i>
    </a>
</li>





<li>
    <a href="https://github.com/ttskch">
        <i class="fa fa-fw fa-github"></i>
    </a>
</li>





<li>
    <a href="https://twitter.com/ttskch">
        <i class="fa fa-fw fa-twitter"></i>
    </a>
</li>
































            </ul>
            <hr class="with-no-margin margin-bottom"/>

            <div class="block sticky">
                <h2>Recent posts</h2>
                <ul>
                    
                    
                    
                    
                    <li><a href="/hisatgenotype/2015/01/01/hello/">Hello, Jekyll and TtskchTheme</a></li>
                    
                </ul>
            </div>
        </div>
    </aside>
</div>

<footer class="site-footer">
    <div class="inner">
        <span>Powered by <a href="http://jekyllrb.com">Jekyll</a> with <a href="https://github.com/ttskch/jekyll-ttskch-theme">TtskchTheme</a></span>
    </div>
</footer>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="/hisatgenotype/assets/lib/garand-sticky/jquery.sticky.js"></script>
<script src="/hisatgenotype/assets/js/script.js"></script>





<!-- Init Facebook SDK -->


<!-- Google Analytics -->



</body>
</html>
