<!DOCTYPE html>
<html lang="en">
<head>
    

    




    

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <title>HOWTO | HISAT-genotype</title>

    <meta name="description" content="next-generation genomic analysis software platform">

    <link rel="shortcut icon" href="/hisat-genotype/assets/img/ogp.png" type="image/x-icon">
    <link rel="canonical" href="http://localhost:4000/hisat-genotype/howto/">
    <link rel="alternate" type="application/atom+xml" title="HISAT-genotype" href="/hisat-genotype/feed.xml" />

    

    <meta property="og:title" content="HOWTO | HISAT-genotype" />
    <meta property="og:type" content="website" />
    <meta property="og:image" content="//ttskch.github.io/jekyll-ttskch-theme/assets/img/ogp.png" />
    <meta property="og:url" content="http://localhost:4000/hisat-genotype/howto/" />
    <meta property="og:site_name" content="HISAT-genotype" />
    <meta property="fb:admins" content="" />
    <meta property="fb:app_id" content="" />
    <meta property="og:description" content="next-generation genomic analysis software platform" />

    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
    <script src="https://use.fontawesome.com/1f5f360d80.js"></script>
    <link href="//fonts.googleapis.com/css?family=Source+Sans+Pro:400,700,700italic,400italic" rel="stylesheet">

    <link href="/hisat-genotype/assets/css/style.css" rel="stylesheet">
</head>
<body>

<header class="site-header">
    <div class="inner clearfix">
        
        <a href="/hisat-genotype/">
            <img class="avatar" src="/hisat-genotype/assets/img/ogp.png" alt=""/>
        </a>
        
        <h1 class="clearfix">
            <a class="title " href="/hisat-genotype/">HISAT-genotype</a>
            <br><span class="description">next-generation genomic analysis software platform</span>
        </h1>
    </div>
</header>

<div class="site-container">
    <div class="site-content">
        <div class="article-wrapper">
    <article>
        

<header>
    <div class="panel">
        <h1>
            
            HOWTO
            
        </h1>

        <ul class="tags">
            
            
            
        </ul>

        <div class="clearfix">
            <ul class="meta">
                

                
            </ul>
        </div>
    </div>

    

    
</header>

        <section class="post-content">
            <h2 class="no_toc" id="howto">HOWTO</h2>

<ul id="markdown-toc">
  <li><a href="#building-indexes" id="markdown-toc-building-indexes">Building indexes</a>    <ul>
      <li><a href="#prepare-data" id="markdown-toc-prepare-data">Prepare data</a></li>
      <li><a href="#build-hfm-index" id="markdown-toc-build-hfm-index">Build HFM index</a></li>
      <li><a href="#build-hgfm-index-with-snps" id="markdown-toc-build-hgfm-index-with-snps">Build HGFM index with SNPs</a></li>
      <li><a href="#build-hgfm-index-with-transcripts" id="markdown-toc-build-hgfm-index-with-transcripts">Build HGFM index with transcripts</a></li>
      <li><a href="#build-hgfm-index-with-snps-and-transcripts" id="markdown-toc-build-hgfm-index-with-snps-and-transcripts">Build HGFM index with SNPs and transcripts</a></li>
    </ul>
  </li>
</ul>

<h3 id="building-indexes">Building indexes</h3>
<p>Depend on your purpose, you have to download reference sequence, gene annotation and SNP files.<br />
We also provides scripts to build indexes. [Download Link]</p>

<h4 id="prepare-data">Prepare data</h4>
<ol>
  <li>Download reference
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ wget ftp://ftp.ensembl.org/pub/release-84/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
$ gzip -d Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
$ mv Homo_sapiens.GRCh38.dna.primary_assembly.fa genome.fa
</code></pre></div>    </div>
  </li>
  <li>Download GTF and make exon, splicesite file.<br />
If you want to build HFM index, you can skip this step.
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ wget ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.gtf.gz  
$ gzip -d Homo_sapiens.GRCh38.84.gtf.gz
$ mv Homo_sapiens.GRCh38.84.gtf genome.gtf
$ hisat2_extract_splice_sites.py genome.gtf &gt; genome.ss
$ hisat2_extract_exons.py genome.gtf &gt; genome.exon
</code></pre></div>    </div>
  </li>
  <li>Download SNP<br />
If you want to build HFM index, you can skip this step.
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ wget http://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/snp144Common.txt.gz
$ gzip -d snp144Common.txt.gz
</code></pre></div>    </div>

    <p>Convert chromosome names of UCSC Database to Ensembl Annotation</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ awk 'BEGIN{OFS="\t"} {if($2 ~ /^chr/) {$2 = substr($2, 4)}; if($2 == "M") {$2 = "MT"} print}' snp144Common.txt &gt; snp144Common.txt.ensembl
</code></pre></div>    </div>

    <p>make SNPs and haplotype file</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ hisat2_extract_snps_haplotypes_UCSC.py genome.fa snp144Common.txt.ensembl genome
</code></pre></div>    </div>
  </li>
</ol>

<h4 id="build-hfm-index">Build HFM index</h4>
<p>It takes about 20 minutes(depend on HW spec) to build index, and requires at least 6GB memory.</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ hisat2-build -p 16 genome.fa genome
</code></pre></div></div>

<h4 id="build-hgfm-index-with-snps">Build HGFM index with SNPs</h4>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ hisat2-build -p 16 --snp genome.snp --haplotype genome.haplotype genome.fa genome_snp
</code></pre></div></div>

<h4 id="build-hgfm-index-with-transcripts">Build HGFM index with transcripts</h4>
<p>It takes about 1 hour(depend on HW spec) to build index, and requires at least 160GB memory.</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ hisat2-build -p 16 --exon genome.exon --ss genome.ss genome.fa genome_tran
</code></pre></div></div>

<h4 id="build-hgfm-index-with-snps-and-transcripts">Build HGFM index with SNPs and transcripts</h4>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ hisat2-build -p 16 --snp genome.snp --haplotype genome.haplotype --exon genome.exon --ss genome.ss genome.fa genome_snp_tran
</code></pre></div></div>


        </section>
        <footer>
    
</footer>

    </article>
</div>

    </div>

    <aside class="site-aside">
        <div class="inner">
            <div class="block">
                <form action="/hisat-genotype/search">
                    <input type="search" id="search" name="q" placeholder="Search" />
                </form>
            </div>
            <ul>
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                
                <li><a class="page-link" href="/hisat-genotype/main/">Main</a></li>
                
                
                
                <li><a class="page-link" href="/hisat-genotype/about/">About</a></li>
                
                
                
                <li><a class="page-link" href="/hisat-genotype/tags/">Tags</a></li>
                
                
                
                <li><a class="page-link" href="/hisat-genotype/archives/">Archives</a></li>
                
                
                
                <li><a class="page-link" href="/hisat-genotype/manual/">Manual</a></li>
                
                
                
                <li><a class="page-link" href="/hisat-genotype/howto/">HOWTO</a></li>
                
                
            </ul>
            <ul class="icons">
                


<li>
    <a href="/hisat-genotype/feed.xml">
        <i class="fa fa-fw fa-rss"></i>
    </a>
</li>





<li>
    <a href="https://github.com/ttskch">
        <i class="fa fa-fw fa-github"></i>
    </a>
</li>





<li>
    <a href="https://twitter.com/ttskch">
        <i class="fa fa-fw fa-twitter"></i>
    </a>
</li>
































            </ul>
            <hr class="with-no-margin margin-bottom"/>

            <div class="block sticky">
                <h2>Recent posts</h2>
                <ul>
                    
                    
                    
                    
                    <li><a href="/hisat-genotype/2015/01/01/hello/">Hello, Jekyll and TtskchTheme</a></li>
                    
                </ul>
            </div>
        </div>
    </aside>
</div>

<footer class="site-footer">
    <div class="inner">
        <span>Powered by <a href="http://jekyllrb.com">Jekyll</a> with <a href="https://github.com/ttskch/jekyll-ttskch-theme">TtskchTheme</a></span>
    </div>
</footer>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="/hisat-genotype/assets/lib/garand-sticky/jquery.sticky.js"></script>
<script src="/hisat-genotype/assets/js/script.js"></script>





<!-- Init Facebook SDK -->


<!-- Google Analytics -->



</body>
</html>
